// @ng-maps/core/esm2015/lib/services/managers/marker.manager.js
import { __awaiter } from "tslib";
import { Injectable, NgZone } from '@angular/core';
import { MapsApiWrapper } from '../maps-api-wrapper';
export class MarkerManager {
    constructor(_mapsWrapper, _zone) {
        this._mapsWrapper = _mapsWrapper;
        this._zone = _zone;
        this._markers = new Map();
    }
    addMarker(marker) {
        return __awaiter(this, void 0, void 0, function* () {
            const m = yield this._mapsWrapper.createMarker({ lat: marker.latitude, lng: marker.longitude }, {
                label: marker.label,
                draggable: marker.draggable,
                icon: marker.iconUrl,
                opacity: marker.opacity,
                optimized: marker.optimized,
                visible: marker.visible,
                zIndex: marker.zIndex,
                title: marker.title,
                clickable: marker.clickable,
                animation: typeof marker.animation === 'string'
                    ? google.maps.Animation[marker.animation]
                    : marker.animation,
            });
            this._markers.set(marker, m);
        });
    }
    getNativeMarker(marker) {
        return this._markers.get(marker);
    }
}
MarkerManager.decorators = [
    { type: Injectable }
];
MarkerManager.ctorParameters = () => [
    { type: MapsApiWrapper },
    { type: NgZone }
];
