// @ng-maps/core/esm2015/lib/services/script-loader.service.js
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class ScriptLoaderService {
    constructor(document) {
        this.alreadyLoaded = new Map();
        this.document = document;
        this.head = this.document.head;
    }
    observeLoad(element) {
        return new Promise((resolve, reject) => {
            function successHandler(event) {
                const readyState = element.readyState;
                // For IE we have readyState, other browsers just call the load event and we proccede
                if (readyState === 'complete' ||
                    readyState === 'loaded' ||
                    event.type === 'load') {
                    resolve(event);
                }
            }
            function errorHandler(event) {
                reject(event);
            }
            element.addEventListener('readystatechange', successHandler);
            element.addEventListener('load', successHandler);
            element.addEventListener('error', errorHandler);
        });
    }
    createScriptElement(src) {
        const script = document.createElement('script');
        script.src = src;
        script.type = 'text/javascript';
        script.async = true;
        script.defer = true;
        return script;
    }
    // tslint:disable-next-line:naming-convention
    createCSSElement(href) {
        const style = document.createElement('link');
        style.rel = 'stylesheet';
        style.type = 'text/css';
        style.href = href;
        return style;
    }
    /**
     * Dynamically loads the given script
     * @param src The url of the script to load dynamically
     * @param integrity set integrity check value
     * @returns Observable<UIEvent> Observable that will be resolved once the script has been loaded.
     */
    loadScript(src, integrity) {
        if (this.alreadyLoaded.has(src)) {
            return this.alreadyLoaded.get(src);
        }
        else {
            const script = this.createScriptElement(src);
            const promise = this.observeLoad(script);
            this.head.appendChild(script);
            this.alreadyLoaded.set(src, promise);
            return promise;
        }
    }
    /**
     * Dynamically loads the given CSS file
     * @param href The url of the CSS to load dynamically
     * @returns Observable<UIEvent> Promise that will be resolved once the CSS file has been loaded.
     */
    loadCSS(href) {
        // tslint:disable-line:naming-convention
        if (this.alreadyLoaded.has(href)) {
            return this.alreadyLoaded.get(href);
        }
        else {
            const style = this.createCSSElement(href);
            const promise = this.observeLoad(style);
            this.head.appendChild(style);
            this.alreadyLoaded.set(href, promise);
            return promise;
        }
    }
}
ScriptLoaderService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ScriptLoaderService_Factory() { return new ScriptLoaderService(i0.ɵɵinject(i1.DOCUMENT)); }, token: ScriptLoaderService, providedIn: "root" });
ScriptLoaderService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
ScriptLoaderService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
