// @ng-maps/core/esm2015/lib/core.module.js
import { InjectionToken, NgModule } from '@angular/core';
import { NgMapsCircleDirective } from './directives/circle';
import { NgMapsFitBoundsDirective } from './directives/fit-bounds';
import { NgMapsInfoWindowComponent } from './directives/info-window';
import { NgMapsViewComponent } from './directives/map';
import { NgMapsMarkerComponent } from './directives/marker';
import { NgMapsPolygon } from './directives/polygon';
import { NgMapsPolyline } from './directives/polyline';
import { NgMapsPolylinePoint } from './directives/polyline-point';
import { NgMapsRectangleDirective } from './directives/rectangle';
export const MAP_PROVIDER = new InjectionToken('Map Provider');
export function isGoogleMaps(token) {
    return token === 'GoogleMaps';
}
export function isHereMaps(token) {
    return token === 'HereMaps';
}
/**
 * @internal
 */
export function coreDirectives() {
    return [
        NgMapsViewComponent,
        NgMapsMarkerComponent,
        NgMapsInfoWindowComponent,
        NgMapsFitBoundsDirective,
        NgMapsCircleDirective,
        NgMapsPolygon,
        NgMapsPolyline,
        NgMapsPolylinePoint,
        NgMapsRectangleDirective,
    ];
}
/**
 * The ng-maps core module. Contains all Directives/Services/Pipes
 */
export class NgMapsCoreModule {
}
NgMapsCoreModule.decorators = [
    { type: NgModule, args: [{
                declarations: coreDirectives(),
                exports: coreDirectives(),
            },] }
];
