// @ng-maps/core/esm2015/lib/directives/info-window.js
import { __awaiter } from "tslib";
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild, } from '@angular/core';
import { InfoWindowManager } from '../services/managers/info-window.manager';
let infoWindowId = 0;
/**
 * NgMapsInfoWindowComponent renders a info window inside a {@link NgMapsMarkerComponent} or standalone.
 *
 * ### Example
 * ```typescript
 * import { Component } from '@angular/core';
 *
 * @Component({
 *  selector: 'my-map-cmp',
 *  styles: [`
 *    map-view {
 *      height: 300px;
 *    }
 * `],
 *  template: `
 *    <map-view [latitude]="lat" [longitude]="lng" [zoom]="zoom">
 *      <map-marker [latitude]="lat" [longitude]="lng" [label]="'M'">
 *        <map-info-window [disableAutoPan]="true">
 *          Hi, this is the content of the <strong>info window</strong>
 *        </map-info-window>
 *      </map-marker>
 *    </map-view>
 *  `
 * })
 * ```
 */
export class NgMapsInfoWindowComponent {
    // @todo how to add correct typings?
    constructor(_infoWindowManager, elementRef) {
        this._infoWindowManager = _infoWindowManager;
        this.elementRef = elementRef;
        /**
         * Sets the open state for the InfoWindow. You can also call the open() and close() methods.
         */
        this.isOpen = false;
        /**
         * Emits an event when the info window is closed.
         */
        this.infoWindowClose = new EventEmitter();
        this._infoWindowAddedToManager = false;
        this._id = (infoWindowId++).toString();
    }
    ngOnInit() {
        this._infoWindowManager.addInfoWindow(this).then(() => {
            this._infoWindowAddedToManager = true;
            this._updateOpenState();
            this._registerEventListeners();
        });
    }
    /** @internal */
    ngOnChanges(changes) {
        if (!this._infoWindowAddedToManager) {
            return;
        }
        if ((changes.latitude || changes.longitude) &&
            typeof this.latitude === 'number' &&
            typeof this.longitude === 'number') {
            this._infoWindowManager.setPosition(this);
        }
        if (changes.zIndex) {
            this._infoWindowManager.setZIndex(this);
        }
        if (changes.isOpen) {
            this._updateOpenState();
        }
        this._setInfoWindowOptions(changes);
    }
    _registerEventListeners() {
        this._infoWindowManager
            .createEventObservable('closeclick', this)
            .subscribe(() => {
            this.isOpen = false;
            this.infoWindowClose.emit();
        });
    }
    _updateOpenState() {
        this.isOpen ? this.open() : this.close();
    }
    _setInfoWindowOptions(changes) {
        const options = {};
        const optionKeys = Object.keys(changes).filter((k) => NgMapsInfoWindowComponent._infoWindowOptionsInputs.includes(k));
        optionKeys.forEach((k) => {
            options[k] = changes[k].currentValue;
        });
        this._infoWindowManager.setOptions(this, options);
    }
    /**
     * Opens the info window.
     */
    open(event) {
        return this._infoWindowManager.open(this, event);
    }
    /**
     * Closes the info window.
     */
    close() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._infoWindowManager.close(this);
            return this.infoWindowClose.emit();
        });
    }
    /** @internal */
    id() {
        return this._id;
    }
    /** @internal */
    toString() {
        return `NgMapsInfoWindowComponent-${this._id.toString()}`;
    }
    /** @internal */
    ngOnDestroy() {
        this._infoWindowManager.deleteInfoWindow(this);
    }
}
NgMapsInfoWindowComponent._infoWindowOptionsInputs = [
    'disableAutoPan',
    'maxWidth',
];
NgMapsInfoWindowComponent.decorators = [
    { type: Component, args: [{
                selector: 'agm-info-window, map-info-window',
                template: `
    <div class="info-window-content" #content>
      <ng-content></ng-content>
    </div>
  `
            },] }
];
NgMapsInfoWindowComponent.ctorParameters = () => [
    { type: InfoWindowManager },
    { type: ElementRef }
];
NgMapsInfoWindowComponent.propDecorators = {
    latitude: [{ type: Input }],
    longitude: [{ type: Input }],
    disableAutoPan: [{ type: Input }],
    zIndex: [{ type: Input }],
    maxWidth: [{ type: Input }],
    content: [{ type: ViewChild, args: ['content', { static: true },] }],
    isOpen: [{ type: Input }],
    infoWindowClose: [{ type: Output }]
};
