// @ng-maps/core/esm2015/lib/directives/map.js
import { __awaiter } from "tslib";
import { Component, ElementRef, EventEmitter, Input, NgZone, Output, ViewChild, } from '@angular/core';
import { Subscription } from 'rxjs';
import { FitBoundsService } from '../services/fit-bounds';
import { MapsApiWrapper } from '../services/maps-api-wrapper';
/**
 * NgMapsViewComponent renders a Google Map.
 * **Important note**: To be able see a map in the browser, you have to define a height for the
 * element `map-view`.
 *
 * @example
 * <map-view [latitude]="lat" [longitude]="lng" [zoom]="zoom"></map-view>
 */
export class NgMapsViewComponent {
    constructor(_mapsWrapper, _fitBoundsService, _zone) {
        this._mapsWrapper = _mapsWrapper;
        this._fitBoundsService = _fitBoundsService;
        this._zone = _zone;
        /**
         * The longitude that defines the center of the map.
         */
        this.longitude = 0;
        /**
         * The latitude that defines the center of the map.
         */
        this.latitude = 0;
        /**
         * The zoom level of the map. The default zoom level is 8.
         */
        this.zoom = 8;
        /**
         * Enables/disables if map is draggable.
         */
        this.draggable = true;
        /**
         * Enables/disables zoom and center on double click. Enabled by default.
         */
        this.disableDoubleClickZoom = false;
        /**
         * Enables/disables all default UI of the Google map. Please note: When the map is created, this
         * value cannot get updated.
         */
        this.disableDefaultUI = false;
        /**
         * If false, disables scrollwheel zooming on the map. The scrollwheel is enabled by default.
         */
        this.scrollwheel = true;
        /**
         * If false, prevents the map from being controlled by the keyboard. Keyboard shortcuts are
         * enabled by default.
         */
        this.keyboardShortcuts = true;
        /**
         * The enabled/disabled state of the Zoom control.
         */
        this.zoomControl = true;
        /**
         * Styles to apply to each of the default map types. Note that for Satellite/Hybrid and Terrain
         * modes, these styles will only apply to labels and geometry.
         */
        this.styles = [];
        /**
         * When true and the latitude and/or longitude values changes, the Google Maps panTo method is
         * used to
         * center the map. See: https://developers.google.com/maps/documentation/javascript/reference#Map
         */
        this.usePanning = false;
        /**
         * Sets the viewport to contain the given bounds.
         * If this option to `true`, the bounds get automatically computed from all elements that use the {@link NgMapsFitBounds} directive.
         */
        this.fitBounds = false;
        /**
         * The initial enabled/disabled state of the Scale control. This is disabled by default.
         */
        this.scaleControl = true;
        /**
         * The initial enabled/disabled state of the Map type control.
         */
        this.mapTypeControl = true;
        /**
         * The initial enabled/disabled state of the Pan control.
         */
        this.panControl = false;
        /**
         * The initial enabled/disabled state of the Rotate control.
         */
        this.rotateControl = false;
        /**
         * The initial enabled/disabled state of the Fullscreen control.
         */
        this.fullscreenControl = false;
        /**
         * The map mapTypeId. Defaults to 'roadmap'.
         */
        this.mapTypeId = 'roadmap';
        this._layerInstance = new Map();
        /**
         * When false, map icons are not clickable. A map icon represents a point of interest,
         * also known as a POI. By default map icons are clickable.
         */
        this.clickableIcons = true;
        /**
         * This setting controls how gestures on the map are handled.
         * Allowed values:
         * - 'cooperative' (Two-finger touch gestures pan and zoom the map. One-finger touch gestures are not handled by the map.)
         * - 'greedy'      (All touch gestures pan or zoom the map.)
         * - 'none'        (The map cannot be panned or zoomed by user gestures.)
         * - 'auto'        [default] (Gesture handling is either cooperative or greedy, depending on whether the page is scrollable or not.
         */
        this.gestureHandling = 'auto';
        /**
         * Controls the automatic switching behavior for the angle of incidence of
         * the map. The only allowed values are 0 and 45. The value 0 causes the map
         * to always use a 0° overhead view regardless of the zoom level and
         * viewport. The value 45 causes the tilt angle to automatically switch to
         * 45 whenever 45° imagery is available for the current zoom level and
         * viewport, and switch back to 0 whenever 45° imagery is not available
         * (this is the default behavior). 45° imagery is only available for
         * satellite and hybrid map types, within some locations, and at some zoom
         * levels. Note: getTilt returns the current tilt angle, not the value
         * specified by this option. Because getTilt and this option refer to
         * different things, do not bind() the tilt property; doing so may yield
         * unpredictable effects. (Default of AGM is 0 (disabled). Enable it with value 45.)
         */
        this.tilt = 0;
        this.subscription = new Subscription();
        /**
         * This event emitter gets emitted when the user clicks on the map (but not when they click on a
         * marker or infoWindow).
         */
        this.mapClick = new EventEmitter();
        /**
         * This event emitter gets emitted when the user right-clicks on the map (but not when they click
         * on a marker or infoWindow).
         */
        this.mapRightClick = new EventEmitter();
        /**
         * This event emitter gets emitted when the user double-clicks on the map (but not when they click
         * on a marker or infoWindow).
         */
        this.mapDblClick = new EventEmitter();
        /**
         * This event emitter is fired when the map center changes.
         */
        this.centerChange = new EventEmitter();
        /**
         * This event is fired when the viewport bounds have changed.
         */
        this.boundsChange = new EventEmitter();
        /**
         * This event is fired when the mapTypeId property changes.
         */
        this.mapTypeIdChange = new EventEmitter();
        /**
         * This event is fired when the map becomes idle after panning or zooming.
         */
        this.idle = new EventEmitter();
        /**
         * This event is fired when the zoom level has changed.
         */
        this.zoomChange = new EventEmitter();
        /**
         * This event is fired when the google map is fully initialized.
         * You get the google.maps.Map instance as a result of this EventEmitter.
         */
        this.mapReady = new EventEmitter();
        /**
         * This event is fired when the visible tiles have finished loading.
         */
        this.tilesLoaded = new EventEmitter();
    }
    /** @internal */
    ngOnInit() {
        this._initMapInstance(this.container.nativeElement);
    }
    _initMapInstance(el) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._mapsWrapper.createMap(el, { lat: this.latitude || 0, lng: this.longitude || 0 }, {
                zoom: this.zoom,
                minZoom: this.minZoom,
                maxZoom: this.maxZoom,
                disableDefaultUI: this.disableDefaultUI,
                disableDoubleClickZoom: this.disableDoubleClickZoom,
                scrollwheel: this.scrollwheel,
                backgroundColor: this.backgroundColor,
                draggable: this.draggable,
                draggableCursor: this.draggableCursor,
                draggingCursor: this.draggingCursor,
                keyboardShortcuts: this.keyboardShortcuts,
                styles: this.styles,
                zoomControl: this.zoomControl,
                zoomControlOptions: this.zoomControlOptions,
                streetViewControl: this.streetViewControl,
                streetViewControlOptions: this.streetViewControlOptions,
                scaleControl: this.scaleControl,
                scaleControlOptions: this.scaleControlOptions,
                mapTypeControl: this.mapTypeControl,
                mapTypeControlOptions: this.mapTypeControlOptions,
                panControl: this.panControl,
                panControlOptions: this.panControlOptions,
                rotateControl: this.rotateControl,
                rotateControlOptions: this.rotateControlOptions,
                fullscreenControl: this.fullscreenControl,
                fullscreenControlOptions: this.fullscreenControlOptions,
                mapTypeId: this.mapTypeId,
                clickableIcons: this.clickableIcons,
                gestureHandling: this.gestureHandling,
                tilt: this.tilt,
            });
            const map = yield this._mapsWrapper.getNativeMap();
            this.mapReady.emit(map);
            // register event listeners
            this._handleMapCenterChange();
            this._handleMapZoomChange();
            this._handleMapMouseEvents();
            this._handleBoundsChange();
            this._handleMapTypeIdChange();
            this._handleTilesLoadedEvent();
            this._handleIdleEvent();
        });
    }
    _handleIdleEvent() {
        throw new Error('Method not implemented.');
    }
    _handleTilesLoadedEvent() {
        throw new Error('Method not implemented.');
    }
    _handleMapTypeIdChange() {
        throw new Error('Method not implemented.');
    }
    _handleBoundsChange() {
        throw new Error('Method not implemented.');
    }
    _handleMapMouseEvents() {
        throw new Error('Method not implemented.');
    }
    _handleMapZoomChange() {
        throw new Error('Method not implemented.');
    }
    _handleMapCenterChange() {
        throw new Error('Method not implemented.');
    }
    /** @internal */
    ngOnDestroy() {
        // unsubscribe all registered observable subscriptions
        this.subscription.unsubscribe();
        // remove all listeners from the map instance
        this._mapsWrapper.clearInstanceListeners();
        if (this._fitBoundsSubscription) {
            this._fitBoundsSubscription.unsubscribe();
        }
    }
    /* @internal */
    ngOnChanges(changes) {
        this._updateMapOptionsChanges(changes);
        this._updatePosition(changes);
        this._layerChanges(changes);
    }
    _updateMapOptionsChanges(changes) {
        const options = {};
        const optionKeys = Object.keys(changes).filter((k) => NgMapsViewComponent._mapOptionsAttributes.includes(k));
        optionKeys.forEach((k) => {
            options[k] = changes[k].currentValue;
        });
        return this._mapsWrapper.setMapOptions(options);
    }
    /**
     * @todo google specific
     * @param changes
     * @protected
     */
    _layerChanges(changes) {
        return __awaiter(this, void 0, void 0, function* () {
            if (changes.layers) {
                const map = yield this._mapsWrapper.getNativeMap();
                const layers = Array.isArray(this.layers) ? this.layers : [this.layers];
                layers.forEach((layer) => {
                    if (!this._layerInstance.has(layer)) {
                        const i = new google.maps[layer]();
                        // @todo typings
                        i.setMap(map);
                        this._layerInstance.set(layer, i);
                    }
                });
                Array.from(this._layerInstance.keys()).forEach((layer) => {
                    if (!layers.includes(layer)) {
                        const i = this._layerInstance.get(layer);
                        i.setMap(null);
                        this._layerInstance.delete(layer);
                    }
                });
            }
        });
    }
    /**
     * Triggers a resize event on the google map instance.
     * When recenter is true, the of the google map gets called with the current lat/lng values or fitBounds value to recenter the map.
     * Returns a promise that gets resolved after the event was triggered.
     */
    triggerResize(recenter = true) {
        // Note: When we would trigger the resize event and show the map in the same turn (which is a
        // common case for triggering a resize event), then the resize event would not
        // work (to show the map), so we trigger the event in a timeout.
        return new Promise((resolve) => {
            setTimeout(() => __awaiter(this, void 0, void 0, function* () {
                yield this._mapsWrapper.triggerMapEvent('resize');
                if (recenter) {
                    this.fitBounds != null ? this._fitBounds() : this._setCenter();
                }
                resolve();
            }));
        });
    }
    _updatePosition(changes) {
        return __awaiter(this, void 0, void 0, function* () {
            if (changes.latitude == null &&
                changes.longitude == null &&
                !changes.fitBounds) {
                // no position update needed
                return;
            }
            // we prefer fitBounds in changes
            if ('fitBounds' in changes) {
                yield this._fitBounds();
                return;
            }
            if (typeof this.latitude === 'string') {
                this.latitude = parseFloat(this.latitude);
            }
            if (typeof this.longitude === 'string') {
                this.longitude = parseFloat(this.longitude);
            }
            const center = yield this._mapsWrapper.getCenter();
            if (!(typeof this.latitude !== 'number' || typeof this.longitude !== 'number') &&
                this.latitude !== center.lat &&
                this.longitude !== center.lng) {
                yield this._setCenter();
                return;
            }
        });
    }
    _setCenter() {
        const newCenter = {
            lat: this.latitude,
            lng: this.longitude,
        };
        if (this.usePanning) {
            return this._mapsWrapper.panTo(newCenter);
        }
        else {
            return this._mapsWrapper.setCenter(newCenter);
        }
    }
    _fitBounds() {
        switch (this.fitBounds) {
            case true:
                this._subscribeToFitBoundsUpdates();
                break;
            case false:
                if (this._fitBoundsSubscription) {
                    this._fitBoundsSubscription.unsubscribe();
                }
                break;
            default:
                if (this._fitBoundsSubscription) {
                    this._fitBoundsSubscription.unsubscribe();
                }
                return this._updateBounds(this.fitBounds);
        }
    }
    _subscribeToFitBoundsUpdates() {
        this._zone.runOutsideAngular(() => {
            this._fitBoundsSubscription = this._fitBoundsService
                .getBounds$()
                .subscribe((b) => {
                return this._zone.run(() => this._updateBounds(b));
            });
        });
    }
    _updateBounds(bounds) {
        return __awaiter(this, void 0, void 0, function* () {
            if (bounds != null) {
                /**
                 * await map to not update bounds till map is initialized
                 */
                yield this._mapsWrapper.getNativeMap();
                if (this.usePanning) {
                    return this._mapsWrapper.panToBounds(bounds, this.boundsPadding);
                }
                else {
                    return this._mapsWrapper.fitBounds(bounds, this.boundsPadding);
                }
            }
        });
    }
}
/**
 * Map option attributes that can change over time
 */
NgMapsViewComponent._mapOptionsAttributes = [
    'disableDoubleClickZoom',
    'scrollwheel',
    'draggable',
    'draggableCursor',
    'draggingCursor',
    'keyboardShortcuts',
    'zoomControl',
    'zoomControlOptions',
    'styles',
    'streetViewControl',
    'streetViewControlOptions',
    'zoom',
    'mapTypeControl',
    'mapTypeControlOptions',
    'minZoom',
    'maxZoom',
    'panControl',
    'panControlOptions',
    'rotateControl',
    'rotateControlOptions',
    'fullscreenControl',
    'fullscreenControlOptions',
    'scaleControl',
    'scaleControlOptions',
    'mapTypeId',
    'clickableIcons',
    'gestureHandling',
    'tilt',
];
NgMapsViewComponent.decorators = [
    { type: Component, args: [{
                template: `
    <div class="map-container-inner" #container></div>
    <div class="map-content">
      <ng-content></ng-content>
    </div>
  `,
                styles: [`
      .map-container-inner {
        width: inherit;
        height: inherit;
      }

      .map-content {
        display: none;
      }
    `]
            },] }
];
NgMapsViewComponent.ctorParameters = () => [
    { type: MapsApiWrapper },
    { type: FitBoundsService },
    { type: NgZone }
];
NgMapsViewComponent.propDecorators = {
    longitude: [{ type: Input }],
    latitude: [{ type: Input }],
    zoom: [{ type: Input }],
    minZoom: [{ type: Input }],
    maxZoom: [{ type: Input }],
    draggable: [{ type: Input }],
    disableDoubleClickZoom: [{ type: Input }],
    disableDefaultUI: [{ type: Input }],
    scrollwheel: [{ type: Input }],
    backgroundColor: [{ type: Input }],
    draggableCursor: [{ type: Input }],
    draggingCursor: [{ type: Input }],
    keyboardShortcuts: [{ type: Input }],
    zoomControl: [{ type: Input }],
    zoomControlOptions: [{ type: Input }],
    styles: [{ type: Input }],
    usePanning: [{ type: Input }],
    streetViewControl: [{ type: Input }],
    streetViewControlOptions: [{ type: Input }],
    fitBounds: [{ type: Input }],
    boundsPadding: [{ type: Input }],
    scaleControl: [{ type: Input }],
    scaleControlOptions: [{ type: Input }],
    mapTypeControl: [{ type: Input }],
    mapTypeControlOptions: [{ type: Input }],
    panControl: [{ type: Input }],
    panControlOptions: [{ type: Input }],
    rotateControl: [{ type: Input }],
    rotateControlOptions: [{ type: Input }],
    fullscreenControl: [{ type: Input }],
    fullscreenControlOptions: [{ type: Input }],
    mapTypeId: [{ type: Input }],
    layers: [{ type: Input }],
    clickableIcons: [{ type: Input }],
    gestureHandling: [{ type: Input }],
    tilt: [{ type: Input }],
    mapClick: [{ type: Output }],
    mapRightClick: [{ type: Output }],
    mapDblClick: [{ type: Output }],
    centerChange: [{ type: Output }],
    boundsChange: [{ type: Output }],
    mapTypeIdChange: [{ type: Output }],
    idle: [{ type: Output }],
    zoomChange: [{ type: Output }],
    mapReady: [{ type: Output }],
    tilesLoaded: [{ type: Output }],
    container: [{ type: ViewChild, args: ['container', { static: true },] }]
};
