// @ng-maps/core/esm2015/lib/directives/marker.js
import { Component, ContentChildren, EventEmitter, forwardRef, Input, Output, QueryList, } from '@angular/core';
import { ReplaySubject, Subscription } from 'rxjs';
import { FitBoundsAccessor } from '../services/fit-bounds';
import { MarkerManager } from '../services/managers/marker.manager';
import { NgMapsInfoWindowComponent } from './info-window';
let markerId = 0;
/**
 * NgMapsMarkerComponent renders a map marker inside a {@link NgMapsViewComponent}.
 *
 * @example
 * <agm-map [latitude]="lat" [longitude]="lng" [zoom]="zoom">
 *               <agm-marker [latitude]="lat" [longitude]="lng" label="M"></agm-marker>
 * </agm-map>
 */
export class NgMapsMarkerComponent {
    constructor(markerManager) {
        this.markerManager = markerManager;
        /**
         * If true, the marker can be dragged. Default value is false.
         */
        // tslint:disable-next-line:no-input-rename
        this.draggable = false;
        /**
         * If true, the marker is visible
         */
        this.visible = true;
        /**
         * Whether to automatically open the child info window when the marker is clicked.
         */
        this.openInfoWindow = true;
        /**
         * The marker's opacity between 0.0 and 1.0.
         */
        this.opacity = 1;
        /**
         * Marker optimize flag. If it is false then it prevent duplicate rendering.
         * Default it is true
         */
        this.optimized = true;
        /**
         * All markers are displayed on the map in order of their zIndex, with higher values displaying in
         * front of markers with lower values. By default, markers are displayed according to their
         * vertical position on screen, with lower markers appearing in front of markers further up the
         * screen.
         */
        this.zIndex = 1;
        /**
         * If true, the marker can be clicked. Default value is true.
         */
        // tslint:disable-next-line:no-input-rename
        this.clickable = true;
        /**
         * This event emitter gets emitted when the user clicks on the marker.
         */
        this.markerClick = new EventEmitter();
        /**
         * This event is fired when the user rightclicks on the marker.
         */
        this.markerRightClick = new EventEmitter();
        /**
         * This event is fired when the user starts dragging the marker.
         */
        this.dragStart = new EventEmitter();
        /**
         * This event is repeatedly fired while the user drags the marker.
         */
        this.drag = new EventEmitter();
        /**
         * This event is fired when the user stops dragging the marker.
         */
        this.dragEnd = new EventEmitter();
        /**
         * This event is fired when the user mouses over the marker.
         */
        this.mouseOver = new EventEmitter();
        /**
         * This event is fired when the user mouses outside the marker.
         */
        this.mouseOut = new EventEmitter();
        /**
         * @internal
         */
        this.infoWindow = new QueryList();
        this._markerAddedToManger = false;
        this.subscription = new Subscription();
        this._fitBoundsDetails$ = new ReplaySubject(1);
        this._id = (markerId++).toString();
    }
    /**
     *  @internal
     */
    ngAfterContentInit() {
        this.handleInfoWindowUpdate();
        this.infoWindow.changes.subscribe(() => this.handleInfoWindowUpdate());
    }
    handleInfoWindowUpdate() {
        if (this.infoWindow.length > 1) {
            throw new Error('Expected no more than one info window.');
        }
        this.infoWindow.forEach((marker) => {
            marker.hostMarker = this;
        });
    }
    /**
     * @internal
     */
    ngOnChanges(changes) {
        if (typeof this.latitude === 'string') {
            this.latitude = Number(this.latitude);
        }
        if (typeof this.longitude === 'string') {
            this.longitude = Number(this.longitude);
        }
        if (typeof this.latitude !== 'number' ||
            typeof this.longitude !== 'number') {
            return;
        }
        if (!this._markerAddedToManger) {
            this.markerManager.addMarker(this).then(() => {
                this._updateFitBoundsDetails();
                this._markerAddedToManger = true;
                this._addEventListeners();
            });
            return;
        }
        if (changes.latitude || changes.longitude) {
            this.markerManager.updateMarkerPosition(this);
            this._updateFitBoundsDetails();
        }
        if (changes.title) {
            this.markerManager.updateTitle(this);
        }
        if (changes.label) {
            this.markerManager.updateLabel(this);
        }
        if (changes.draggable) {
            this.markerManager.updateDraggable(this);
        }
        if (changes.iconUrl) {
            this.markerManager.updateIconLegacy(this);
        }
        if (changes.icon) {
            this.markerManager.updateIcon(this);
        }
        if (changes.opacity) {
            this.markerManager.updateOpacity(this);
        }
        if (changes.visible) {
            this.markerManager.updateVisible(this);
        }
        if (changes.zIndex) {
            this.markerManager.updateZIndex(this);
        }
        if (changes.clickable) {
            this.markerManager.updateClickable(this);
        }
        if (changes.animation) {
            this.markerManager.updateAnimation(this);
        }
    }
    /**
     * @internal
     */
    getFitBoundsDetails$() {
        return this._fitBoundsDetails$.asObservable();
    }
    _updateFitBoundsDetails() {
        this._fitBoundsDetails$.next({
            latLng: { lat: this.latitude, lng: this.longitude },
        });
    }
    _addEventListeners() {
        const cs = this.markerManager
            .createEventObservable(['click', 'pointerdown'], this)
            .subscribe({
            next: (event) => {
                if (this.openInfoWindow) {
                    this.infoWindow.forEach((infoWindow) => infoWindow.open(event));
                }
                this.markerClick.emit(this);
            },
        });
        this.subscription.add(cs);
        const rc = this.markerManager
            .createEventObservable('rightclick', this)
            .subscribe(() => {
            this.markerRightClick.emit(null);
        });
        this.subscription.add(rc);
        const ds = this.markerManager
            .createEventObservable('dragstart', this)
            .subscribe((e) => {
            this.dragStart.emit(e);
        });
        this.subscription.add(ds);
        const d = this.markerManager
            .createEventObservable('drag', this)
            .subscribe((e) => {
            this.drag.emit(e);
        });
        this.subscription.add(d);
        const dragend = this.markerManager
            .createEventObservable('dragend', this)
            .subscribe((e) => {
            this.dragEnd.emit(e);
        });
        this.subscription.add(dragend);
        const mouseover = this.markerManager
            .createEventObservable(['mouseover', 'pointerenter'], this)
            .subscribe((e) => {
            this.mouseOver.emit(e);
        });
        this.subscription.add(mouseover);
        const mouseout = this.markerManager
            .createEventObservable(['mouseout', 'pointerleave'], this)
            .subscribe((e) => {
            this.mouseOut.emit(e);
        });
        this.subscription.add(mouseout);
    }
    /** @internal */
    id() {
        return this._id;
    }
    /** @internal */
    toString() {
        return `NgMapsMarker-${this._id}`;
    }
    /** @internal */
    ngOnDestroy() {
        this.markerManager.deleteMarker(this);
        // unsubscribe all registered observable subscription
        this.subscription.unsubscribe();
    }
}
NgMapsMarkerComponent.decorators = [
    { type: Component, args: [{
                selector: 'map-marker',
                providers: [
                    {
                        provide: FitBoundsAccessor,
                        useExisting: forwardRef(() => NgMapsMarkerComponent),
                    },
                ],
                inputs: [
                    'latitude',
                    'longitude',
                    'title',
                    'label',
                    'draggable: markerDraggable',
                    'iconUrl',
                    'icon',
                    'openInfoWindow',
                    'opacity',
                    'optimized',
                    'visible',
                    'zIndex',
                    'animation',
                ],
                outputs: [
                    'markerClick',
                    'dragStart',
                    'drag',
                    'dragEnd',
                    'mouseOver',
                    'mouseOut',
                ],
                template: '<ng-content></ng-content>'
            },] }
];
NgMapsMarkerComponent.ctorParameters = () => [
    { type: MarkerManager }
];
NgMapsMarkerComponent.propDecorators = {
    latitude: [{ type: Input }],
    longitude: [{ type: Input }],
    title: [{ type: Input }],
    label: [{ type: Input }],
    draggable: [{ type: Input, args: ['markerDraggable',] }],
    iconUrl: [{ type: Input }],
    icon: [{ type: Input }],
    visible: [{ type: Input }],
    openInfoWindow: [{ type: Input }],
    opacity: [{ type: Input }],
    optimized: [{ type: Input }],
    zIndex: [{ type: Input }],
    clickable: [{ type: Input, args: ['markerClickable',] }],
    markerClick: [{ type: Output }],
    markerRightClick: [{ type: Output }],
    dragStart: [{ type: Output }],
    drag: [{ type: Output }],
    dragEnd: [{ type: Output }],
    mouseOver: [{ type: Output }],
    mouseOut: [{ type: Output }],
    infoWindow: [{ type: ContentChildren, args: [NgMapsInfoWindowComponent,] }]
};
