// @ng-maps/core/esm2015/lib/directives/polyline-point.js
import { Directive, EventEmitter, Input, Output, } from '@angular/core';
/**
 * NgMapsPolylinePoint represents one element of a polyline within a  {@link
 * NgMapsPolyline}
 */
export class NgMapsPolylinePoint {
    constructor() {
        /**
         * This event emitter gets emitted when the position of the point changed.
         */
        this.positionChanged = new EventEmitter();
    }
    ngOnChanges(changes) {
        if (changes.latitude || changes.longitude) {
            const position = {
                lat: changes.latitude ? changes.latitude.currentValue : this.latitude,
                lng: changes.longitude
                    ? changes.longitude.currentValue
                    : this.longitude,
            };
            this.positionChanged.emit(position);
        }
    }
}
NgMapsPolylinePoint.decorators = [
    { type: Directive, args: [{
                selector: 'map-polyline-point',
            },] }
];
NgMapsPolylinePoint.ctorParameters = () => [];
NgMapsPolylinePoint.propDecorators = {
    latitude: [{ type: Input }],
    longitude: [{ type: Input }],
    positionChanged: [{ type: Output }]
};
