// @ng-maps/core/esm2015/lib/directives/polyline.js
import { ContentChildren, Directive, EventEmitter, Input, Output, QueryList, } from '@angular/core';
import { Subscription } from 'rxjs';
import { NgMapsPolylinePoint } from './polyline-point';
import { PolylineManager } from '../services/managers/polyline-manager';
let polylineId = 0;
/**
 * NgMapsPolyline renders a polyline on a {@link https://ng-maps.github.io/core/components/NgMapsViewComponent.html|NgMapsView}
 */
export class NgMapsPolyline {
    constructor(_polylineManager) {
        this._polylineManager = _polylineManager;
        /**
         * Indicates whether this Polyline handles mouse events. Defaults to true.
         */
        this.clickable = true;
        /**
         * If set to true, the user can drag this shape over the map. The geodesic property defines the
         * mode of dragging. Defaults to false.
         */
        // tslint:disable-next-line:no-input-rename
        this.draggable = false;
        /**
         * If set to true, the user can edit this shape by dragging the control points shown at the
         * vertices and on each segment. Defaults to false.
         */
        this.editable = false;
        /**
         * When true, edges of the polygon are interpreted as geodesic and will follow the curvature of
         * the Earth. When false, edges of the polygon are rendered as straight lines in screen space.
         * Note that the shape of a geodesic polygon may appear to change when dragged, as the dimensions
         * are maintained relative to the surface of the earth. Defaults to false.
         */
        this.geodesic = false;
        /**
         * Whether this polyline is visible on the map. Defaults to true.
         */
        this.visible = true;
        /**
         * This event is fired when the DOM click event is fired on the Polyline.
         */
        this.lineClick = new EventEmitter();
        /**
         * This event is fired when the DOM dblclick event is fired on the Polyline.
         */
        this.lineDblClick = new EventEmitter();
        /**
         * This event is repeatedly fired while the user drags the polyline.
         */
        this.lineDrag = new EventEmitter();
        /**
         * This event is fired when the user stops dragging the polyline.
         */
        this.lineDragEnd = new EventEmitter();
        /**
         * This event is fired when the user starts dragging the polyline.
         */
        this.lineDragStart = new EventEmitter();
        /**
         * This event is fired when the DOM mousedown event is fired on the Polyline.
         */
        this.lineMouseDown = new EventEmitter();
        /**
         * This event is fired when the DOM mousemove event is fired on the Polyline.
         */
        this.lineMouseMove = new EventEmitter();
        /**
         * This event is fired on Polyline mouseout.
         */
        this.lineMouseOut = new EventEmitter();
        /**
         * This event is fired on Polyline mouseover.
         */
        this.lineMouseOver = new EventEmitter();
        /**
         * This event is fired whe the DOM mouseup event is fired on the Polyline
         */
        this.lineMouseUp = new EventEmitter();
        /**
         * This even is fired when the Polyline is right-clicked on.
         */
        this.lineRightClick = new EventEmitter();
        this._polylineAddedToManager = false;
        this.subscription = new Subscription();
        this._id = (polylineId++).toString();
    }
    /** @internal */
    ngAfterContentInit() {
        if (this.points.length) {
            this.points.forEach((point) => {
                const subscription = point.positionChanged.subscribe(() => {
                    this._polylineManager.updatePolylinePoints(this);
                });
                this.subscription.add(subscription);
            });
        }
        if (!this._polylineAddedToManager) {
            this._init();
        }
        const s = this.points.changes.subscribe(() => this._polylineManager.updatePolylinePoints(this));
        this.subscription.add(s);
        this._polylineManager.updatePolylinePoints(this);
    }
    ngOnChanges(changes) {
        if (!this._polylineAddedToManager) {
            this._init();
            return;
        }
        const options = {};
        const optionKeys = Object.keys(changes).filter((k) => NgMapsPolyline._polylineOptionsAttributes.indexOf(k) !== -1);
        optionKeys.forEach((k) => (options[k] = changes[k].currentValue));
        this._polylineManager.setPolylineOptions(this, options);
    }
    _init() {
        this._polylineManager.addPolyline(this);
        this._polylineAddedToManager = true;
        this._addEventListeners();
    }
    _addEventListeners() {
        const handlers = [
            {
                name: 'click',
                handler: (ev) => this.lineClick.emit(ev),
            },
            {
                name: 'dblclick',
                handler: (ev) => this.lineDblClick.emit(ev),
            },
            { name: 'drag', handler: (ev) => this.lineDrag.emit(ev) },
            {
                name: 'dragend',
                handler: (ev) => this.lineDragEnd.emit(ev),
            },
            {
                name: 'dragstart',
                handler: (ev) => this.lineDragStart.emit(ev),
            },
            {
                name: 'mousedown',
                handler: (ev) => this.lineMouseDown.emit(ev),
            },
            {
                name: 'mousemove',
                handler: (ev) => this.lineMouseMove.emit(ev),
            },
            {
                name: 'mouseout',
                handler: (ev) => this.lineMouseOut.emit(ev),
            },
            {
                name: 'mouseover',
                handler: (ev) => this.lineMouseOver.emit(ev),
            },
            {
                name: 'mouseup',
                handler: (ev) => this.lineMouseUp.emit(ev),
            },
            {
                name: 'rightclick',
                handler: (ev) => this.lineRightClick.emit(ev),
            },
        ];
        handlers.forEach((obj) => {
            const os = this._polylineManager
                .createEventObservable(obj.name, this)
                .subscribe(obj.handler);
            this.subscription.add(os);
        });
    }
    /** @internal */
    _getPoints() {
        if (this.points) {
            return this.points.toArray();
        }
        else {
            return [];
        }
    }
    /** @internal */
    id() {
        return this._id;
    }
    /** @internal */
    ngOnDestroy() {
        this._polylineManager.deletePolyline(this);
        // unsubscribe all registered observable subscriptions
        this.subscription.unsubscribe();
    }
}
NgMapsPolyline._polylineOptionsAttributes = [
    'draggable',
    'editable',
    'visible',
    'geodesic',
    'strokeColor',
    'strokeOpacity',
    'strokeWeight',
    'zIndex',
    'icons',
];
NgMapsPolyline.decorators = [
    { type: Directive, args: [{
                selector: 'map-polyline',
            },] }
];
NgMapsPolyline.ctorParameters = () => [
    { type: PolylineManager }
];
NgMapsPolyline.propDecorators = {
    clickable: [{ type: Input }],
    draggable: [{ type: Input, args: ['polylineDraggable',] }],
    editable: [{ type: Input }],
    geodesic: [{ type: Input }],
    strokeColor: [{ type: Input }],
    strokeOpacity: [{ type: Input }],
    strokeWeight: [{ type: Input }],
    icons: [{ type: Input }],
    visible: [{ type: Input }],
    zIndex: [{ type: Input }],
    lineClick: [{ type: Output }],
    lineDblClick: [{ type: Output }],
    lineDrag: [{ type: Output }],
    lineDragEnd: [{ type: Output }],
    lineDragStart: [{ type: Output }],
    lineMouseDown: [{ type: Output }],
    lineMouseMove: [{ type: Output }],
    lineMouseOut: [{ type: Output }],
    lineMouseOver: [{ type: Output }],
    lineMouseUp: [{ type: Output }],
    lineRightClick: [{ type: Output }],
    points: [{ type: ContentChildren, args: [NgMapsPolylinePoint,] }]
};
