// @ng-maps/core/bundles/ng-maps-core.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('rxjs/operators'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@ng-maps/core', ['exports', '@angular/core', 'rxjs', 'rxjs/operators', '@angular/common'], factory) :
    (global = global || self, factory((global['ng-maps'] = global['ng-maps'] || {}, global['ng-maps'].core = {}), global.ng.core, global.rxjs, global.rxjs.operators, global.ng.common));
}(this, (function (exports, i0, rxjs, operators, i1) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var MapsAPILoader = /** @class */ (function () {
        function MapsAPILoader() {
        }
        return MapsAPILoader;
    }());
    MapsAPILoader.decorators = [
        { type: i0.Injectable }
    ];

    /**
     * Class to implement when you what to be able to make it work with the auto fit bounds feature
     * of AGM.
     */
    var FitBoundsAccessor = /** @class */ (function () {
        function FitBoundsAccessor() {
        }
        return FitBoundsAccessor;
    }());
    /**
     * The FitBoundsService is responsible for computing the bounds of the a single map.
     */
    var FitBoundsService = /** @class */ (function () {
        function FitBoundsService(loader) {
            var _this = this;
            this._boundsChangeSampleTime$ = new rxjs.BehaviorSubject(200);
            this._includeInBounds$ = new rxjs.BehaviorSubject(new Map());
            this.bounds$ = rxjs.from(loader.load()).pipe(operators.flatMap(function () { return _this._includeInBounds$; }), operators.sample(this._boundsChangeSampleTime$.pipe(operators.switchMap(function (time) { return rxjs.timer(0, time); }))), operators.map(function (includeInBounds) { return _this.generateBounds(includeInBounds); }), operators.shareReplay(1));
        }
        FitBoundsService.prototype.addToBounds = function (latLng) {
            var id = this._createIdentifier(latLng);
            if (this._includeInBounds$.value.has(id)) {
                return;
            }
            var bounds = this._includeInBounds$.value;
            bounds.set(id, latLng);
            this._includeInBounds$.next(bounds);
        };
        FitBoundsService.prototype.removeFromBounds = function (latLng) {
            var bounds = this._includeInBounds$.value;
            bounds.delete(this._createIdentifier(latLng));
            this._includeInBounds$.next(bounds);
        };
        FitBoundsService.prototype.changeFitBoundsChangeSampleTime = function (timeMs) {
            this._boundsChangeSampleTime$.next(timeMs);
        };
        FitBoundsService.prototype.getBounds$ = function () {
            return this.bounds$;
        };
        FitBoundsService.prototype._createIdentifier = function (latLng) {
            return latLng.lat + "+" + latLng.lng;
        };
        return FitBoundsService;
    }());
    FitBoundsService.decorators = [
        { type: i0.Injectable }
    ];
    FitBoundsService.ctorParameters = function () { return [
        { type: MapsAPILoader }
    ]; };

    var MapsApiWrapper = /** @class */ (function () {
        function MapsApiWrapper(_loader, _zone) {
            var _this = this;
            this._loader = _loader;
            this._zone = _zone;
            this._api = new Promise(function (resolve) {
                _this._mapResolver = resolve;
            });
        }
        MapsApiWrapper.prototype.getNativeMap = function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    return [2 /*return*/, this._api];
                });
            });
        };
        return MapsApiWrapper;
    }());
    MapsApiWrapper.decorators = [
        { type: i0.Injectable }
    ];
    MapsApiWrapper.ctorParameters = function () { return [
        { type: MapsAPILoader },
        { type: i0.NgZone }
    ]; };

    /**
     * NgMapsViewComponent renders a Google Map.
     * **Important note**: To be able see a map in the browser, you have to define a height for the
     * element `map-view`.
     *
     * @example
     * <map-view [latitude]="lat" [longitude]="lng" [zoom]="zoom"></map-view>
     */
    var NgMapsViewComponent = /** @class */ (function () {
        function NgMapsViewComponent(_mapsWrapper, _fitBoundsService, _zone) {
            this._mapsWrapper = _mapsWrapper;
            this._fitBoundsService = _fitBoundsService;
            this._zone = _zone;
            /**
             * The longitude that defines the center of the map.
             */
            this.longitude = 0;
            /**
             * The latitude that defines the center of the map.
             */
            this.latitude = 0;
            /**
             * The zoom level of the map. The default zoom level is 8.
             */
            this.zoom = 8;
            /**
             * Enables/disables if map is draggable.
             */
            this.draggable = true;
            /**
             * Enables/disables zoom and center on double click. Enabled by default.
             */
            this.disableDoubleClickZoom = false;
            /**
             * Enables/disables all default UI of the Google map. Please note: When the map is created, this
             * value cannot get updated.
             */
            this.disableDefaultUI = false;
            /**
             * If false, disables scrollwheel zooming on the map. The scrollwheel is enabled by default.
             */
            this.scrollwheel = true;
            /**
             * If false, prevents the map from being controlled by the keyboard. Keyboard shortcuts are
             * enabled by default.
             */
            this.keyboardShortcuts = true;
            /**
             * The enabled/disabled state of the Zoom control.
             */
            this.zoomControl = true;
            /**
             * Styles to apply to each of the default map types. Note that for Satellite/Hybrid and Terrain
             * modes, these styles will only apply to labels and geometry.
             */
            this.styles = [];
            /**
             * When true and the latitude and/or longitude values changes, the Google Maps panTo method is
             * used to
             * center the map. See: https://developers.google.com/maps/documentation/javascript/reference#Map
             */
            this.usePanning = false;
            /**
             * Sets the viewport to contain the given bounds.
             * If this option to `true`, the bounds get automatically computed from all elements that use the {@link NgMapsFitBounds} directive.
             */
            this.fitBounds = false;
            /**
             * The initial enabled/disabled state of the Scale control. This is disabled by default.
             */
            this.scaleControl = true;
            /**
             * The initial enabled/disabled state of the Map type control.
             */
            this.mapTypeControl = true;
            /**
             * The initial enabled/disabled state of the Pan control.
             */
            this.panControl = false;
            /**
             * The initial enabled/disabled state of the Rotate control.
             */
            this.rotateControl = false;
            /**
             * The initial enabled/disabled state of the Fullscreen control.
             */
            this.fullscreenControl = false;
            /**
             * The map mapTypeId. Defaults to 'roadmap'.
             */
            this.mapTypeId = 'roadmap';
            this._layerInstance = new Map();
            /**
             * When false, map icons are not clickable. A map icon represents a point of interest,
             * also known as a POI. By default map icons are clickable.
             */
            this.clickableIcons = true;
            /**
             * This setting controls how gestures on the map are handled.
             * Allowed values:
             * - 'cooperative' (Two-finger touch gestures pan and zoom the map. One-finger touch gestures are not handled by the map.)
             * - 'greedy'      (All touch gestures pan or zoom the map.)
             * - 'none'        (The map cannot be panned or zoomed by user gestures.)
             * - 'auto'        [default] (Gesture handling is either cooperative or greedy, depending on whether the page is scrollable or not.
             */
            this.gestureHandling = 'auto';
            /**
             * Controls the automatic switching behavior for the angle of incidence of
             * the map. The only allowed values are 0 and 45. The value 0 causes the map
             * to always use a 0° overhead view regardless of the zoom level and
             * viewport. The value 45 causes the tilt angle to automatically switch to
             * 45 whenever 45° imagery is available for the current zoom level and
             * viewport, and switch back to 0 whenever 45° imagery is not available
             * (this is the default behavior). 45° imagery is only available for
             * satellite and hybrid map types, within some locations, and at some zoom
             * levels. Note: getTilt returns the current tilt angle, not the value
             * specified by this option. Because getTilt and this option refer to
             * different things, do not bind() the tilt property; doing so may yield
             * unpredictable effects. (Default of AGM is 0 (disabled). Enable it with value 45.)
             */
            this.tilt = 0;
            this.subscription = new rxjs.Subscription();
            /**
             * This event emitter gets emitted when the user clicks on the map (but not when they click on a
             * marker or infoWindow).
             */
            this.mapClick = new i0.EventEmitter();
            /**
             * This event emitter gets emitted when the user right-clicks on the map (but not when they click
             * on a marker or infoWindow).
             */
            this.mapRightClick = new i0.EventEmitter();
            /**
             * This event emitter gets emitted when the user double-clicks on the map (but not when they click
             * on a marker or infoWindow).
             */
            this.mapDblClick = new i0.EventEmitter();
            /**
             * This event emitter is fired when the map center changes.
             */
            this.centerChange = new i0.EventEmitter();
            /**
             * This event is fired when the viewport bounds have changed.
             */
            this.boundsChange = new i0.EventEmitter();
            /**
             * This event is fired when the mapTypeId property changes.
             */
            this.mapTypeIdChange = new i0.EventEmitter();
            /**
             * This event is fired when the map becomes idle after panning or zooming.
             */
            this.idle = new i0.EventEmitter();
            /**
             * This event is fired when the zoom level has changed.
             */
            this.zoomChange = new i0.EventEmitter();
            /**
             * This event is fired when the google map is fully initialized.
             * You get the google.maps.Map instance as a result of this EventEmitter.
             */
            this.mapReady = new i0.EventEmitter();
            /**
             * This event is fired when the visible tiles have finished loading.
             */
            this.tilesLoaded = new i0.EventEmitter();
        }
        /** @internal */
        NgMapsViewComponent.prototype.ngOnInit = function () {
            this._initMapInstance(this.container.nativeElement);
        };
        NgMapsViewComponent.prototype._initMapInstance = function (el) {
            return __awaiter(this, void 0, void 0, function () {
                var map;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, this._mapsWrapper.createMap(el, { lat: this.latitude || 0, lng: this.longitude || 0 }, {
                                zoom: this.zoom,
                                minZoom: this.minZoom,
                                maxZoom: this.maxZoom,
                                disableDefaultUI: this.disableDefaultUI,
                                disableDoubleClickZoom: this.disableDoubleClickZoom,
                                scrollwheel: this.scrollwheel,
                                backgroundColor: this.backgroundColor,
                                draggable: this.draggable,
                                draggableCursor: this.draggableCursor,
                                draggingCursor: this.draggingCursor,
                                keyboardShortcuts: this.keyboardShortcuts,
                                styles: this.styles,
                                zoomControl: this.zoomControl,
                                zoomControlOptions: this.zoomControlOptions,
                                streetViewControl: this.streetViewControl,
                                streetViewControlOptions: this.streetViewControlOptions,
                                scaleControl: this.scaleControl,
                                scaleControlOptions: this.scaleControlOptions,
                                mapTypeControl: this.mapTypeControl,
                                mapTypeControlOptions: this.mapTypeControlOptions,
                                panControl: this.panControl,
                                panControlOptions: this.panControlOptions,
                                rotateControl: this.rotateControl,
                                rotateControlOptions: this.rotateControlOptions,
                                fullscreenControl: this.fullscreenControl,
                                fullscreenControlOptions: this.fullscreenControlOptions,
                                mapTypeId: this.mapTypeId,
                                clickableIcons: this.clickableIcons,
                                gestureHandling: this.gestureHandling,
                                tilt: this.tilt,
                            })];
                        case 1:
                            _a.sent();
                            return [4 /*yield*/, this._mapsWrapper.getNativeMap()];
                        case 2:
                            map = _a.sent();
                            this.mapReady.emit(map);
                            // register event listeners
                            this._handleMapCenterChange();
                            this._handleMapZoomChange();
                            this._handleMapMouseEvents();
                            this._handleBoundsChange();
                            this._handleMapTypeIdChange();
                            this._handleTilesLoadedEvent();
                            this._handleIdleEvent();
                            return [2 /*return*/];
                    }
                });
            });
        };
        NgMapsViewComponent.prototype._handleIdleEvent = function () {
            throw new Error('Method not implemented.');
        };
        NgMapsViewComponent.prototype._handleTilesLoadedEvent = function () {
            throw new Error('Method not implemented.');
        };
        NgMapsViewComponent.prototype._handleMapTypeIdChange = function () {
            throw new Error('Method not implemented.');
        };
        NgMapsViewComponent.prototype._handleBoundsChange = function () {
            throw new Error('Method not implemented.');
        };
        NgMapsViewComponent.prototype._handleMapMouseEvents = function () {
            throw new Error('Method not implemented.');
        };
        NgMapsViewComponent.prototype._handleMapZoomChange = function () {
            throw new Error('Method not implemented.');
        };
        NgMapsViewComponent.prototype._handleMapCenterChange = function () {
            throw new Error('Method not implemented.');
        };
        /** @internal */
        NgMapsViewComponent.prototype.ngOnDestroy = function () {
            // unsubscribe all registered observable subscriptions
            this.subscription.unsubscribe();
            // remove all listeners from the map instance
            this._mapsWrapper.clearInstanceListeners();
            if (this._fitBoundsSubscription) {
                this._fitBoundsSubscription.unsubscribe();
            }
        };
        /* @internal */
        NgMapsViewComponent.prototype.ngOnChanges = function (changes) {
            this._updateMapOptionsChanges(changes);
            this._updatePosition(changes);
            this._layerChanges(changes);
        };
        NgMapsViewComponent.prototype._updateMapOptionsChanges = function (changes) {
            var options = {};
            var optionKeys = Object.keys(changes).filter(function (k) { return NgMapsViewComponent._mapOptionsAttributes.includes(k); });
            optionKeys.forEach(function (k) {
                options[k] = changes[k].currentValue;
            });
            return this._mapsWrapper.setMapOptions(options);
        };
        /**
         * @todo google specific
         * @param changes
         * @protected
         */
        NgMapsViewComponent.prototype._layerChanges = function (changes) {
            return __awaiter(this, void 0, void 0, function () {
                var map_1, layers_1;
                var _this = this;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!changes.layers) return [3 /*break*/, 2];
                            return [4 /*yield*/, this._mapsWrapper.getNativeMap()];
                        case 1:
                            map_1 = _a.sent();
                            layers_1 = Array.isArray(this.layers) ? this.layers : [this.layers];
                            layers_1.forEach(function (layer) {
                                if (!_this._layerInstance.has(layer)) {
                                    var i = new google.maps[layer]();
                                    // @todo typings
                                    i.setMap(map_1);
                                    _this._layerInstance.set(layer, i);
                                }
                            });
                            Array.from(this._layerInstance.keys()).forEach(function (layer) {
                                if (!layers_1.includes(layer)) {
                                    var i = _this._layerInstance.get(layer);
                                    i.setMap(null);
                                    _this._layerInstance.delete(layer);
                                }
                            });
                            _a.label = 2;
                        case 2: return [2 /*return*/];
                    }
                });
            });
        };
        /**
         * Triggers a resize event on the google map instance.
         * When recenter is true, the of the google map gets called with the current lat/lng values or fitBounds value to recenter the map.
         * Returns a promise that gets resolved after the event was triggered.
         */
        NgMapsViewComponent.prototype.triggerResize = function (recenter) {
            var _this = this;
            if (recenter === void 0) { recenter = true; }
            // Note: When we would trigger the resize event and show the map in the same turn (which is a
            // common case for triggering a resize event), then the resize event would not
            // work (to show the map), so we trigger the event in a timeout.
            return new Promise(function (resolve) {
                setTimeout(function () { return __awaiter(_this, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0: return [4 /*yield*/, this._mapsWrapper.triggerMapEvent('resize')];
                            case 1:
                                _a.sent();
                                if (recenter) {
                                    this.fitBounds != null ? this._fitBounds() : this._setCenter();
                                }
                                resolve();
                                return [2 /*return*/];
                        }
                    });
                }); });
            });
        };
        NgMapsViewComponent.prototype._updatePosition = function (changes) {
            return __awaiter(this, void 0, void 0, function () {
                var center;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (changes.latitude == null &&
                                changes.longitude == null &&
                                !changes.fitBounds) {
                                // no position update needed
                                return [2 /*return*/];
                            }
                            if (!('fitBounds' in changes)) return [3 /*break*/, 2];
                            return [4 /*yield*/, this._fitBounds()];
                        case 1:
                            _a.sent();
                            return [2 /*return*/];
                        case 2:
                            if (typeof this.latitude === 'string') {
                                this.latitude = parseFloat(this.latitude);
                            }
                            if (typeof this.longitude === 'string') {
                                this.longitude = parseFloat(this.longitude);
                            }
                            return [4 /*yield*/, this._mapsWrapper.getCenter()];
                        case 3:
                            center = _a.sent();
                            if (!(!(typeof this.latitude !== 'number' || typeof this.longitude !== 'number') &&
                                this.latitude !== center.lat &&
                                this.longitude !== center.lng)) return [3 /*break*/, 5];
                            return [4 /*yield*/, this._setCenter()];
                        case 4:
                            _a.sent();
                            return [2 /*return*/];
                        case 5: return [2 /*return*/];
                    }
                });
            });
        };
        NgMapsViewComponent.prototype._setCenter = function () {
            var newCenter = {
                lat: this.latitude,
                lng: this.longitude,
            };
            if (this.usePanning) {
                return this._mapsWrapper.panTo(newCenter);
            }
            else {
                return this._mapsWrapper.setCenter(newCenter);
            }
        };
        NgMapsViewComponent.prototype._fitBounds = function () {
            switch (this.fitBounds) {
                case true:
                    this._subscribeToFitBoundsUpdates();
                    break;
                case false:
                    if (this._fitBoundsSubscription) {
                        this._fitBoundsSubscription.unsubscribe();
                    }
                    break;
                default:
                    if (this._fitBoundsSubscription) {
                        this._fitBoundsSubscription.unsubscribe();
                    }
                    return this._updateBounds(this.fitBounds);
            }
        };
        NgMapsViewComponent.prototype._subscribeToFitBoundsUpdates = function () {
            var _this = this;
            this._zone.runOutsideAngular(function () {
                _this._fitBoundsSubscription = _this._fitBoundsService
                    .getBounds$()
                    .subscribe(function (b) {
                    return _this._zone.run(function () { return _this._updateBounds(b); });
                });
            });
        };
        NgMapsViewComponent.prototype._updateBounds = function (bounds) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!(bounds != null)) return [3 /*break*/, 2];
                            /**
                             * await map to not update bounds till map is initialized
                             */
                            return [4 /*yield*/, this._mapsWrapper.getNativeMap()];
                        case 1:
                            /**
                             * await map to not update bounds till map is initialized
                             */
                            _a.sent();
                            if (this.usePanning) {
                                return [2 /*return*/, this._mapsWrapper.panToBounds(bounds, this.boundsPadding)];
                            }
                            else {
                                return [2 /*return*/, this._mapsWrapper.fitBounds(bounds, this.boundsPadding)];
                            }
                            _a.label = 2;
                        case 2: return [2 /*return*/];
                    }
                });
            });
        };
        return NgMapsViewComponent;
    }());
    /**
     * Map option attributes that can change over time
     */
    NgMapsViewComponent._mapOptionsAttributes = [
        'disableDoubleClickZoom',
        'scrollwheel',
        'draggable',
        'draggableCursor',
        'draggingCursor',
        'keyboardShortcuts',
        'zoomControl',
        'zoomControlOptions',
        'styles',
        'streetViewControl',
        'streetViewControlOptions',
        'zoom',
        'mapTypeControl',
        'mapTypeControlOptions',
        'minZoom',
        'maxZoom',
        'panControl',
        'panControlOptions',
        'rotateControl',
        'rotateControlOptions',
        'fullscreenControl',
        'fullscreenControlOptions',
        'scaleControl',
        'scaleControlOptions',
        'mapTypeId',
        'clickableIcons',
        'gestureHandling',
        'tilt',
    ];
    NgMapsViewComponent.decorators = [
        { type: i0.Component, args: [{
                    template: "\n    <div class=\"map-container-inner\" #container></div>\n    <div class=\"map-content\">\n      <ng-content></ng-content>\n    </div>\n  ",
                    styles: ["\n      .map-container-inner {\n        width: inherit;\n        height: inherit;\n      }\n\n      .map-content {\n        display: none;\n      }\n    "]
                },] }
    ];
    NgMapsViewComponent.ctorParameters = function () { return [
        { type: MapsApiWrapper },
        { type: FitBoundsService },
        { type: i0.NgZone }
    ]; };
    NgMapsViewComponent.propDecorators = {
        longitude: [{ type: i0.Input }],
        latitude: [{ type: i0.Input }],
        zoom: [{ type: i0.Input }],
        minZoom: [{ type: i0.Input }],
        maxZoom: [{ type: i0.Input }],
        draggable: [{ type: i0.Input }],
        disableDoubleClickZoom: [{ type: i0.Input }],
        disableDefaultUI: [{ type: i0.Input }],
        scrollwheel: [{ type: i0.Input }],
        backgroundColor: [{ type: i0.Input }],
        draggableCursor: [{ type: i0.Input }],
        draggingCursor: [{ type: i0.Input }],
        keyboardShortcuts: [{ type: i0.Input }],
        zoomControl: [{ type: i0.Input }],
        zoomControlOptions: [{ type: i0.Input }],
        styles: [{ type: i0.Input }],
        usePanning: [{ type: i0.Input }],
        streetViewControl: [{ type: i0.Input }],
        streetViewControlOptions: [{ type: i0.Input }],
        fitBounds: [{ type: i0.Input }],
        boundsPadding: [{ type: i0.Input }],
        scaleControl: [{ type: i0.Input }],
        scaleControlOptions: [{ type: i0.Input }],
        mapTypeControl: [{ type: i0.Input }],
        mapTypeControlOptions: [{ type: i0.Input }],
        panControl: [{ type: i0.Input }],
        panControlOptions: [{ type: i0.Input }],
        rotateControl: [{ type: i0.Input }],
        rotateControlOptions: [{ type: i0.Input }],
        fullscreenControl: [{ type: i0.Input }],
        fullscreenControlOptions: [{ type: i0.Input }],
        mapTypeId: [{ type: i0.Input }],
        layers: [{ type: i0.Input }],
        clickableIcons: [{ type: i0.Input }],
        gestureHandling: [{ type: i0.Input }],
        tilt: [{ type: i0.Input }],
        mapClick: [{ type: i0.Output }],
        mapRightClick: [{ type: i0.Output }],
        mapDblClick: [{ type: i0.Output }],
        centerChange: [{ type: i0.Output }],
        boundsChange: [{ type: i0.Output }],
        mapTypeIdChange: [{ type: i0.Output }],
        idle: [{ type: i0.Output }],
        zoomChange: [{ type: i0.Output }],
        mapReady: [{ type: i0.Output }],
        tilesLoaded: [{ type: i0.Output }],
        container: [{ type: i0.ViewChild, args: ['container', { static: true },] }]
    };

    var MarkerManager = /** @class */ (function () {
        function MarkerManager(_mapsWrapper, _zone) {
            this._mapsWrapper = _mapsWrapper;
            this._zone = _zone;
            this._markers = new Map();
        }
        MarkerManager.prototype.addMarker = function (marker) {
            return __awaiter(this, void 0, void 0, function () {
                var m;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, this._mapsWrapper.createMarker({ lat: marker.latitude, lng: marker.longitude }, {
                                label: marker.label,
                                draggable: marker.draggable,
                                icon: marker.iconUrl,
                                opacity: marker.opacity,
                                optimized: marker.optimized,
                                visible: marker.visible,
                                zIndex: marker.zIndex,
                                title: marker.title,
                                clickable: marker.clickable,
                                animation: typeof marker.animation === 'string'
                                    ? google.maps.Animation[marker.animation]
                                    : marker.animation,
                            })];
                        case 1:
                            m = _a.sent();
                            this._markers.set(marker, m);
                            return [2 /*return*/];
                    }
                });
            });
        };
        MarkerManager.prototype.getNativeMarker = function (marker) {
            return this._markers.get(marker);
        };
        return MarkerManager;
    }());
    MarkerManager.decorators = [
        { type: i0.Injectable }
    ];
    MarkerManager.ctorParameters = function () { return [
        { type: MapsApiWrapper },
        { type: i0.NgZone }
    ]; };

    var InfoWindowManager = /** @class */ (function () {
        function InfoWindowManager(_mapsWrapper, _zone, _markerManager) {
            this._mapsWrapper = _mapsWrapper;
            this._zone = _zone;
            this._markerManager = _markerManager;
            this._infoWindows = new Map();
        }
        return InfoWindowManager;
    }());
    InfoWindowManager.decorators = [
        { type: i0.Injectable }
    ];
    InfoWindowManager.ctorParameters = function () { return [
        { type: MapsApiWrapper },
        { type: i0.NgZone },
        { type: MarkerManager }
    ]; };

    var infoWindowId = 0;
    /**
     * NgMapsInfoWindowComponent renders a info window inside a {@link NgMapsMarkerComponent} or standalone.
     *
     * ### Example
     * ```typescript
     * import { Component } from '@angular/core';
     *
     * @Component({
     *  selector: 'my-map-cmp',
     *  styles: [`
     *    map-view {
     *      height: 300px;
     *    }
     * `],
     *  template: `
     *    <map-view [latitude]="lat" [longitude]="lng" [zoom]="zoom">
     *      <map-marker [latitude]="lat" [longitude]="lng" [label]="'M'">
     *        <map-info-window [disableAutoPan]="true">
     *          Hi, this is the content of the <strong>info window</strong>
     *        </map-info-window>
     *      </map-marker>
     *    </map-view>
     *  `
     * })
     * ```
     */
    var NgMapsInfoWindowComponent = /** @class */ (function () {
        // @todo how to add correct typings?
        function NgMapsInfoWindowComponent(_infoWindowManager, elementRef) {
            this._infoWindowManager = _infoWindowManager;
            this.elementRef = elementRef;
            /**
             * Sets the open state for the InfoWindow. You can also call the open() and close() methods.
             */
            this.isOpen = false;
            /**
             * Emits an event when the info window is closed.
             */
            this.infoWindowClose = new i0.EventEmitter();
            this._infoWindowAddedToManager = false;
            this._id = (infoWindowId++).toString();
        }
        NgMapsInfoWindowComponent.prototype.ngOnInit = function () {
            var _this = this;
            this._infoWindowManager.addInfoWindow(this).then(function () {
                _this._infoWindowAddedToManager = true;
                _this._updateOpenState();
                _this._registerEventListeners();
            });
        };
        /** @internal */
        NgMapsInfoWindowComponent.prototype.ngOnChanges = function (changes) {
            if (!this._infoWindowAddedToManager) {
                return;
            }
            if ((changes.latitude || changes.longitude) &&
                typeof this.latitude === 'number' &&
                typeof this.longitude === 'number') {
                this._infoWindowManager.setPosition(this);
            }
            if (changes.zIndex) {
                this._infoWindowManager.setZIndex(this);
            }
            if (changes.isOpen) {
                this._updateOpenState();
            }
            this._setInfoWindowOptions(changes);
        };
        NgMapsInfoWindowComponent.prototype._registerEventListeners = function () {
            var _this = this;
            this._infoWindowManager
                .createEventObservable('closeclick', this)
                .subscribe(function () {
                _this.isOpen = false;
                _this.infoWindowClose.emit();
            });
        };
        NgMapsInfoWindowComponent.prototype._updateOpenState = function () {
            this.isOpen ? this.open() : this.close();
        };
        NgMapsInfoWindowComponent.prototype._setInfoWindowOptions = function (changes) {
            var options = {};
            var optionKeys = Object.keys(changes).filter(function (k) { return NgMapsInfoWindowComponent._infoWindowOptionsInputs.includes(k); });
            optionKeys.forEach(function (k) {
                options[k] = changes[k].currentValue;
            });
            this._infoWindowManager.setOptions(this, options);
        };
        /**
         * Opens the info window.
         */
        NgMapsInfoWindowComponent.prototype.open = function (event) {
            return this._infoWindowManager.open(this, event);
        };
        /**
         * Closes the info window.
         */
        NgMapsInfoWindowComponent.prototype.close = function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, this._infoWindowManager.close(this)];
                        case 1:
                            _a.sent();
                            return [2 /*return*/, this.infoWindowClose.emit()];
                    }
                });
            });
        };
        /** @internal */
        NgMapsInfoWindowComponent.prototype.id = function () {
            return this._id;
        };
        /** @internal */
        NgMapsInfoWindowComponent.prototype.toString = function () {
            return "NgMapsInfoWindowComponent-" + this._id.toString();
        };
        /** @internal */
        NgMapsInfoWindowComponent.prototype.ngOnDestroy = function () {
            this._infoWindowManager.deleteInfoWindow(this);
        };
        return NgMapsInfoWindowComponent;
    }());
    NgMapsInfoWindowComponent._infoWindowOptionsInputs = [
        'disableAutoPan',
        'maxWidth',
    ];
    NgMapsInfoWindowComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'agm-info-window, map-info-window',
                    template: "\n    <div class=\"info-window-content\" #content>\n      <ng-content></ng-content>\n    </div>\n  "
                },] }
    ];
    NgMapsInfoWindowComponent.ctorParameters = function () { return [
        { type: InfoWindowManager },
        { type: i0.ElementRef }
    ]; };
    NgMapsInfoWindowComponent.propDecorators = {
        latitude: [{ type: i0.Input }],
        longitude: [{ type: i0.Input }],
        disableAutoPan: [{ type: i0.Input }],
        zIndex: [{ type: i0.Input }],
        maxWidth: [{ type: i0.Input }],
        content: [{ type: i0.ViewChild, args: ['content', { static: true },] }],
        isOpen: [{ type: i0.Input }],
        infoWindowClose: [{ type: i0.Output }]
    };

    var markerId = 0;
    /**
     * NgMapsMarkerComponent renders a map marker inside a {@link NgMapsViewComponent}.
     *
     * @example
     * <agm-map [latitude]="lat" [longitude]="lng" [zoom]="zoom">
     *               <agm-marker [latitude]="lat" [longitude]="lng" label="M"></agm-marker>
     * </agm-map>
     */
    var NgMapsMarkerComponent = /** @class */ (function () {
        function NgMapsMarkerComponent(markerManager) {
            this.markerManager = markerManager;
            /**
             * If true, the marker can be dragged. Default value is false.
             */
            // tslint:disable-next-line:no-input-rename
            this.draggable = false;
            /**
             * If true, the marker is visible
             */
            this.visible = true;
            /**
             * Whether to automatically open the child info window when the marker is clicked.
             */
            this.openInfoWindow = true;
            /**
             * The marker's opacity between 0.0 and 1.0.
             */
            this.opacity = 1;
            /**
             * Marker optimize flag. If it is false then it prevent duplicate rendering.
             * Default it is true
             */
            this.optimized = true;
            /**
             * All markers are displayed on the map in order of their zIndex, with higher values displaying in
             * front of markers with lower values. By default, markers are displayed according to their
             * vertical position on screen, with lower markers appearing in front of markers further up the
             * screen.
             */
            this.zIndex = 1;
            /**
             * If true, the marker can be clicked. Default value is true.
             */
            // tslint:disable-next-line:no-input-rename
            this.clickable = true;
            /**
             * This event emitter gets emitted when the user clicks on the marker.
             */
            this.markerClick = new i0.EventEmitter();
            /**
             * This event is fired when the user rightclicks on the marker.
             */
            this.markerRightClick = new i0.EventEmitter();
            /**
             * This event is fired when the user starts dragging the marker.
             */
            this.dragStart = new i0.EventEmitter();
            /**
             * This event is repeatedly fired while the user drags the marker.
             */
            this.drag = new i0.EventEmitter();
            /**
             * This event is fired when the user stops dragging the marker.
             */
            this.dragEnd = new i0.EventEmitter();
            /**
             * This event is fired when the user mouses over the marker.
             */
            this.mouseOver = new i0.EventEmitter();
            /**
             * This event is fired when the user mouses outside the marker.
             */
            this.mouseOut = new i0.EventEmitter();
            /**
             * @internal
             */
            this.infoWindow = new i0.QueryList();
            this._markerAddedToManger = false;
            this.subscription = new rxjs.Subscription();
            this._fitBoundsDetails$ = new rxjs.ReplaySubject(1);
            this._id = (markerId++).toString();
        }
        /**
         *  @internal
         */
        NgMapsMarkerComponent.prototype.ngAfterContentInit = function () {
            var _this = this;
            this.handleInfoWindowUpdate();
            this.infoWindow.changes.subscribe(function () { return _this.handleInfoWindowUpdate(); });
        };
        NgMapsMarkerComponent.prototype.handleInfoWindowUpdate = function () {
            var _this = this;
            if (this.infoWindow.length > 1) {
                throw new Error('Expected no more than one info window.');
            }
            this.infoWindow.forEach(function (marker) {
                marker.hostMarker = _this;
            });
        };
        /**
         * @internal
         */
        NgMapsMarkerComponent.prototype.ngOnChanges = function (changes) {
            var _this = this;
            if (typeof this.latitude === 'string') {
                this.latitude = Number(this.latitude);
            }
            if (typeof this.longitude === 'string') {
                this.longitude = Number(this.longitude);
            }
            if (typeof this.latitude !== 'number' ||
                typeof this.longitude !== 'number') {
                return;
            }
            if (!this._markerAddedToManger) {
                this.markerManager.addMarker(this).then(function () {
                    _this._updateFitBoundsDetails();
                    _this._markerAddedToManger = true;
                    _this._addEventListeners();
                });
                return;
            }
            if (changes.latitude || changes.longitude) {
                this.markerManager.updateMarkerPosition(this);
                this._updateFitBoundsDetails();
            }
            if (changes.title) {
                this.markerManager.updateTitle(this);
            }
            if (changes.label) {
                this.markerManager.updateLabel(this);
            }
            if (changes.draggable) {
                this.markerManager.updateDraggable(this);
            }
            if (changes.iconUrl) {
                this.markerManager.updateIconLegacy(this);
            }
            if (changes.icon) {
                this.markerManager.updateIcon(this);
            }
            if (changes.opacity) {
                this.markerManager.updateOpacity(this);
            }
            if (changes.visible) {
                this.markerManager.updateVisible(this);
            }
            if (changes.zIndex) {
                this.markerManager.updateZIndex(this);
            }
            if (changes.clickable) {
                this.markerManager.updateClickable(this);
            }
            if (changes.animation) {
                this.markerManager.updateAnimation(this);
            }
        };
        /**
         * @internal
         */
        NgMapsMarkerComponent.prototype.getFitBoundsDetails$ = function () {
            return this._fitBoundsDetails$.asObservable();
        };
        NgMapsMarkerComponent.prototype._updateFitBoundsDetails = function () {
            this._fitBoundsDetails$.next({
                latLng: { lat: this.latitude, lng: this.longitude },
            });
        };
        NgMapsMarkerComponent.prototype._addEventListeners = function () {
            var _this = this;
            var cs = this.markerManager
                .createEventObservable(['click', 'pointerdown'], this)
                .subscribe({
                next: function (event) {
                    if (_this.openInfoWindow) {
                        _this.infoWindow.forEach(function (infoWindow) { return infoWindow.open(event); });
                    }
                    _this.markerClick.emit(_this);
                },
            });
            this.subscription.add(cs);
            var rc = this.markerManager
                .createEventObservable('rightclick', this)
                .subscribe(function () {
                _this.markerRightClick.emit(null);
            });
            this.subscription.add(rc);
            var ds = this.markerManager
                .createEventObservable('dragstart', this)
                .subscribe(function (e) {
                _this.dragStart.emit(e);
            });
            this.subscription.add(ds);
            var d = this.markerManager
                .createEventObservable('drag', this)
                .subscribe(function (e) {
                _this.drag.emit(e);
            });
            this.subscription.add(d);
            var dragend = this.markerManager
                .createEventObservable('dragend', this)
                .subscribe(function (e) {
                _this.dragEnd.emit(e);
            });
            this.subscription.add(dragend);
            var mouseover = this.markerManager
                .createEventObservable(['mouseover', 'pointerenter'], this)
                .subscribe(function (e) {
                _this.mouseOver.emit(e);
            });
            this.subscription.add(mouseover);
            var mouseout = this.markerManager
                .createEventObservable(['mouseout', 'pointerleave'], this)
                .subscribe(function (e) {
                _this.mouseOut.emit(e);
            });
            this.subscription.add(mouseout);
        };
        /** @internal */
        NgMapsMarkerComponent.prototype.id = function () {
            return this._id;
        };
        /** @internal */
        NgMapsMarkerComponent.prototype.toString = function () {
            return "NgMapsMarker-" + this._id;
        };
        /** @internal */
        NgMapsMarkerComponent.prototype.ngOnDestroy = function () {
            this.markerManager.deleteMarker(this);
            // unsubscribe all registered observable subscription
            this.subscription.unsubscribe();
        };
        return NgMapsMarkerComponent;
    }());
    NgMapsMarkerComponent.decorators = [
        { type: i0.Component, args: [{
                    selector: 'map-marker',
                    providers: [
                        {
                            provide: FitBoundsAccessor,
                            useExisting: i0.forwardRef(function () { return NgMapsMarkerComponent; }),
                        },
                    ],
                    inputs: [
                        'latitude',
                        'longitude',
                        'title',
                        'label',
                        'draggable: markerDraggable',
                        'iconUrl',
                        'icon',
                        'openInfoWindow',
                        'opacity',
                        'optimized',
                        'visible',
                        'zIndex',
                        'animation',
                    ],
                    outputs: [
                        'markerClick',
                        'dragStart',
                        'drag',
                        'dragEnd',
                        'mouseOver',
                        'mouseOut',
                    ],
                    template: '<ng-content></ng-content>'
                },] }
    ];
    NgMapsMarkerComponent.ctorParameters = function () { return [
        { type: MarkerManager }
    ]; };
    NgMapsMarkerComponent.propDecorators = {
        latitude: [{ type: i0.Input }],
        longitude: [{ type: i0.Input }],
        title: [{ type: i0.Input }],
        label: [{ type: i0.Input }],
        draggable: [{ type: i0.Input, args: ['markerDraggable',] }],
        iconUrl: [{ type: i0.Input }],
        icon: [{ type: i0.Input }],
        visible: [{ type: i0.Input }],
        openInfoWindow: [{ type: i0.Input }],
        opacity: [{ type: i0.Input }],
        optimized: [{ type: i0.Input }],
        zIndex: [{ type: i0.Input }],
        clickable: [{ type: i0.Input, args: ['markerClickable',] }],
        markerClick: [{ type: i0.Output }],
        markerRightClick: [{ type: i0.Output }],
        dragStart: [{ type: i0.Output }],
        drag: [{ type: i0.Output }],
        dragEnd: [{ type: i0.Output }],
        mouseOver: [{ type: i0.Output }],
        mouseOut: [{ type: i0.Output }],
        infoWindow: [{ type: i0.ContentChildren, args: [NgMapsInfoWindowComponent,] }]
    };

    /**
     * Adds the given directive to the auto fit bounds feature when the value is true.
     * To make it work with you custom AGM component, you also have to implement the {@link FitBoundsAccessor} abstract class.
     * @example
     * <map-marker [mapFitBounds]="true"></map-marker>
     */
    var NgMapsFitBoundsDirective = /** @class */ (function () {
        function NgMapsFitBoundsDirective(_fitBoundsAccessor, _fitBoundsService) {
            this._fitBoundsAccessor = _fitBoundsAccessor;
            this._fitBoundsService = _fitBoundsService;
            /**
             * If the value is true, the element gets added to the bounds of the map.
             * Default: true.
             */
            this.mapFitBounds = true;
            this._destroyed$ = new rxjs.Subject();
            this._latestFitBoundsDetails = null;
        }
        /**
         * @internal
         */
        NgMapsFitBoundsDirective.prototype.ngOnChanges = function (changes) {
            this._updateBounds();
        };
        /**
         * @internal
         */
        NgMapsFitBoundsDirective.prototype.ngOnInit = function () {
            var _this = this;
            this._fitBoundsAccessor
                .getFitBoundsDetails$()
                .pipe(operators.distinctUntilChanged(function (x, y) { return x.latLng.lat === y.latLng.lat && x.latLng.lng === y.latLng.lng; }), operators.takeUntil(this._destroyed$))
                .subscribe(function (details) { return _this._updateBounds(details); });
        };
        /**
         * Either the location changed, or visible status changed.
         * Possible state changes are
         * invisible -> visible
         * visible -> invisible
         * visible -> visible (new location)
         */
        NgMapsFitBoundsDirective.prototype._updateBounds = function (newFitBoundsDetails) {
            // either visibility will change, or location, so remove the old one anyway
            if (this._latestFitBoundsDetails) {
                this._fitBoundsService.removeFromBounds(this._latestFitBoundsDetails.latLng);
                // don't set latestFitBoundsDetails to null, because we can toggle visibility from
                // true -> false -> true, in which case we still need old value cached here
            }
            if (newFitBoundsDetails) {
                this._latestFitBoundsDetails = newFitBoundsDetails;
            }
            if (this._latestFitBoundsDetails) {
                if (this.mapFitBounds === true) {
                    this._fitBoundsService.addToBounds(this._latestFitBoundsDetails.latLng);
                }
                else {
                    this._fitBoundsService.removeFromBounds(this._latestFitBoundsDetails.latLng);
                }
            }
        };
        /**
         * @internal
         */
        NgMapsFitBoundsDirective.prototype.ngOnDestroy = function () {
            this._destroyed$.next();
            this._destroyed$.complete();
            if (this._latestFitBoundsDetails !== null) {
                this._fitBoundsService.removeFromBounds(this._latestFitBoundsDetails.latLng);
            }
        };
        return NgMapsFitBoundsDirective;
    }());
    NgMapsFitBoundsDirective.decorators = [
        { type: i0.Directive, args: [{
                    selector: '[mapFitBounds]',
                },] }
    ];
    NgMapsFitBoundsDirective.ctorParameters = function () { return [
        { type: FitBoundsAccessor, decorators: [{ type: i0.Self }] },
        { type: FitBoundsService }
    ]; };
    NgMapsFitBoundsDirective.propDecorators = {
        mapFitBounds: [{ type: i0.Input }]
    };

    var CircleManager = /** @class */ (function () {
        function CircleManager(_apiWrapper, _zone) {
            this._apiWrapper = _apiWrapper;
            this._zone = _zone;
            this._circles = new Map();
        }
        return CircleManager;
    }());
    CircleManager.decorators = [
        { type: i0.Injectable }
    ];
    CircleManager.ctorParameters = function () { return [
        { type: MapsApiWrapper },
        { type: i0.NgZone }
    ]; };

    var NgMapsCircleDirective = /** @class */ (function () {
        function NgMapsCircleDirective(_manager) {
            this._manager = _manager;
            /**
             * Indicates whether this Circle handles mouse events. Defaults to true.
             */
            this.clickable = true;
            /**
             * If set to true, the user can drag this circle over the map. Defaults to false.
             */
            this.draggable = false;
            /**
             * If set to true, the user can edit this circle by dragging the control points shown at
             * the center and around the circumference of the circle. Defaults to false.
             */
            this.editable = false;
            /**
             * The radius in meters on the Earth's surface.
             */
            this.radius = 0;
            /**
             * The stroke position. Defaults to CENTER.
             * This property is not supported on Internet Explorer 8 and earlier.
             * @fixme
             */
            // @Input() strokePosition: google.maps.StrokePosition =
            //   google.maps.StrokePosition.CENTER;
            /**
             * The stroke width in pixels.
             */
            this.strokeWeight = 0;
            /**
             * Whether this circle is visible on the map. Defaults to true.
             */
            this.visible = true;
            /**
             * This event is fired when the circle's center is changed.
             */
            this.centerChange = new i0.EventEmitter();
            /**
             * This event emitter gets emitted when the user clicks on the circle.
             */
            this.circleClick = new i0.EventEmitter();
            /**
             * This event emitter gets emitted when the user clicks on the circle.
             */
            this.circleDblClick = new i0.EventEmitter();
            /**
             * This event is repeatedly fired while the user drags the circle.
             */
            this.drag = new i0.EventEmitter();
            /**
             * This event is fired when the user stops dragging the circle.
             */
            this.dragEnd = new i0.EventEmitter();
            /**
             * This event is fired when the user starts dragging the circle.
             */
            this.dragStart = new i0.EventEmitter();
            /**
             * This event is fired when the DOM mousedown event is fired on the circle.
             */
            this.mouseDown = new i0.EventEmitter();
            /**
             * This event is fired when the DOM mousemove event is fired on the circle.
             */
            this.mouseMove = new i0.EventEmitter();
            /**
             * This event is fired on circle mouseout.
             */
            this.mouseOut = new i0.EventEmitter();
            /**
             * This event is fired on circle mouseover.
             */
            this.mouseOver = new i0.EventEmitter();
            /**
             * This event is fired when the DOM mouseup event is fired on the circle.
             */
            this.mouseUp = new i0.EventEmitter();
            /**
             * This event is fired when the circle's radius is changed.
             */
            this.radiusChange = new i0.EventEmitter();
            /**
             * This event is fired when the circle is right-clicked on.
             */
            this.rightClick = new i0.EventEmitter();
            this._circleAddedToManager = false;
            this.subscription = new rxjs.Subscription();
        }
        /** @internal */
        NgMapsCircleDirective.prototype.ngOnInit = function () {
            this._manager.addCircle(this);
            this._circleAddedToManager = true;
            this._registerEventListeners();
        };
        /** @internal */
        NgMapsCircleDirective.prototype.ngOnChanges = function (changes) {
            if (!this._circleAddedToManager) {
                return;
            }
            if (changes.latitude || changes.longitude) {
                this._manager.setCenter(this);
            }
            if (changes.editable) {
                this._manager.setEditable(this);
            }
            if (changes.draggable) {
                this._manager.setDraggable(this);
            }
            if (changes.visible) {
                this._manager.setVisible(this);
            }
            if (changes.radius) {
                this._manager.setRadius(this);
            }
            this._updateCircleOptionsChanges(changes);
        };
        NgMapsCircleDirective.prototype._updateCircleOptionsChanges = function (changes) {
            var options = {};
            var optionKeys = Object.keys(changes).filter(function (k) { return NgMapsCircleDirective._mapOptions.indexOf(k) !== -1; });
            optionKeys.forEach(function (k) {
                options[k] = changes[k].currentValue;
            });
            if (optionKeys.length > 0) {
                this._manager.setOptions(this, options);
            }
        };
        NgMapsCircleDirective.prototype._registerEventListeners = function () {
            var _this = this;
            var events = new Map();
            events.set('center_changed', this.centerChange);
            events.set('click', this.circleClick);
            events.set('dblclick', this.circleDblClick);
            events.set('drag', this.drag);
            events.set('dragend', this.dragEnd);
            events.set('dragstart', this.dragStart);
            events.set('mousedown', this.mouseDown);
            events.set('mousemove', this.mouseMove);
            events.set('mouseout', this.mouseOut);
            events.set('mouseover', this.mouseOver);
            events.set('mouseup', this.mouseUp);
            events.set('radius_changed', this.radiusChange);
            events.set('rightclick', this.rightClick);
            events.forEach(function (eventEmitter, eventName) {
                _this.subscription.add(_this._manager
                    .createEventObservable(eventName, _this)
                    .subscribe(function (value) {
                    switch (eventName) {
                        case 'radius_changed':
                            _this._manager
                                .getRadius(_this)
                                .then(function (radius) { return eventEmitter.emit(radius); });
                            break;
                        case 'center_changed':
                            _this._manager.getCenter(_this).then(function (center) { return eventEmitter.emit(Object.assign({}, center)); });
                            break;
                        default:
                            eventEmitter.emit(value);
                    }
                }));
            });
        };
        /** @internal */
        NgMapsCircleDirective.prototype.ngOnDestroy = function () {
            this.subscription.unsubscribe();
            this._manager.removeCircle(this);
        };
        /**
         * Gets the LatLngBounds of this Circle.
         */
        NgMapsCircleDirective.prototype.getBounds = function () {
            return this._manager.getBounds(this);
        };
        NgMapsCircleDirective.prototype.getCenter = function () {
            return this._manager.getCenter(this);
        };
        return NgMapsCircleDirective;
    }());
    NgMapsCircleDirective._mapOptions = [
        'fillColor',
        'fillOpacity',
        'strokeColor',
        'strokeOpacity',
        'strokePosition',
        'strokeWeight',
        'visible',
        'zIndex',
        'clickable',
    ];
    NgMapsCircleDirective.decorators = [
        { type: i0.Directive, args: [{
                    selector: 'map-circle',
                },] }
    ];
    NgMapsCircleDirective.ctorParameters = function () { return [
        { type: CircleManager }
    ]; };
    NgMapsCircleDirective.propDecorators = {
        latitude: [{ type: i0.Input }],
        longitude: [{ type: i0.Input }],
        clickable: [{ type: i0.Input }],
        draggable: [{ type: i0.Input }],
        editable: [{ type: i0.Input }],
        fillColor: [{ type: i0.Input }],
        fillOpacity: [{ type: i0.Input }],
        radius: [{ type: i0.Input }],
        strokeColor: [{ type: i0.Input }],
        strokeOpacity: [{ type: i0.Input }],
        strokeWeight: [{ type: i0.Input }],
        visible: [{ type: i0.Input }],
        zIndex: [{ type: i0.Input }],
        centerChange: [{ type: i0.Output }],
        circleClick: [{ type: i0.Output }],
        circleDblClick: [{ type: i0.Output }],
        drag: [{ type: i0.Output }],
        dragEnd: [{ type: i0.Output }],
        dragStart: [{ type: i0.Output }],
        mouseDown: [{ type: i0.Output }],
        mouseMove: [{ type: i0.Output }],
        mouseOut: [{ type: i0.Output }],
        mouseOver: [{ type: i0.Output }],
        mouseUp: [{ type: i0.Output }],
        radiusChange: [{ type: i0.Output }],
        rightClick: [{ type: i0.Output }]
    };

    var PolygonManager = /** @class */ (function () {
        function PolygonManager(_mapsWrapper, _zone) {
            this._mapsWrapper = _mapsWrapper;
            this._zone = _zone;
            this._polygons = new Map();
        }
        return PolygonManager;
    }());
    PolygonManager.decorators = [
        { type: i0.Injectable }
    ];
    PolygonManager.ctorParameters = function () { return [
        { type: MapsApiWrapper },
        { type: i0.NgZone }
    ]; };

    /**
     * NgMapsPolygon renders a polygon on a {@link https://ng-maps.github.io/core/components/NgMapsViewComponent.html|NgMapsView}
     */
    var NgMapsPolygon = /** @class */ (function () {
        function NgMapsPolygon(_polygonManager) {
            this._polygonManager = _polygonManager;
            /**
             * Indicates whether this Polygon handles mouse events. Defaults to true.
             */
            this.clickable = true;
            /**
             * If set to true, the user can drag this shape over the map. The geodesic
             * property defines the mode of dragging. Defaults to false.
             */
            // tslint:disable-next-line:no-input-rename
            this.draggable = false;
            /**
             * If set to true, the user can edit this shape by dragging the control
             * points shown at the vertices and on each segment. Defaults to false.
             */
            this.editable = false;
            /**
             * When true, edges of the polygon are interpreted as geodesic and will
             * follow the curvature of the Earth. When false, edges of the polygon are
             * rendered as straight lines in screen space. Note that the shape of a
             * geodesic polygon may appear to change when dragged, as the dimensions
             * are maintained relative to the surface of the earth. Defaults to false.
             */
            this.geodesic = false;
            /**
             * The ordered sequence of coordinates that designates a closed loop.
             * Unlike polylines, a polygon may consist of one or more paths.
             *  As a result, the paths property may specify one or more arrays of
             * LatLng coordinates. Paths are closed automatically; do not repeat the
             * first vertex of the path as the last vertex. Simple polygons may be
             * defined using a single array of LatLngs. More complex polygons may
             * specify an array of arrays. Any simple arrays are converted into Arrays.
             * Inserting or removing LatLngs from the Array will automatically update
             * the polygon on the map.
             */
            this.paths = [];
            /**
             * This event is fired when the DOM click event is fired on the Polygon.
             */
            this.polyClick = new i0.EventEmitter();
            /**
             * This event is fired when the DOM dblclick event is fired on the Polygon.
             */
            this.polyDblClick = new i0.EventEmitter();
            /**
             * This event is repeatedly fired while the user drags the polygon.
             */
            this.polyDrag = new i0.EventEmitter();
            /**
             * This event is fired when the user stops dragging the polygon.
             */
            this.polyDragEnd = new i0.EventEmitter();
            /**
             * This event is fired when the user starts dragging the polygon.
             */
            this.polyDragStart = new i0.EventEmitter();
            /**
             * This event is fired when the DOM mousedown event is fired on the Polygon.
             */
            this.polyMouseDown = new i0.EventEmitter();
            /**
             * This event is fired when the DOM mousemove event is fired on the Polygon.
             */
            this.polyMouseMove = new i0.EventEmitter();
            /**
             * This event is fired on Polygon mouseout.
             */
            this.polyMouseOut = new i0.EventEmitter();
            /**
             * This event is fired on Polygon mouseover.
             */
            this.polyMouseOver = new i0.EventEmitter();
            /**
             * This event is fired whe the DOM mouseup event is fired on the Polygon
             */
            this.polyMouseUp = new i0.EventEmitter();
            /**
             * This even is fired when the Polygon is right-clicked on.
             */
            this.polyRightClick = new i0.EventEmitter();
            this._polygonAddedToManager = false;
            this.subscription = new rxjs.Subscription();
        }
        /** @internal */
        NgMapsPolygon.prototype.ngAfterContentInit = function () {
            if (!this._polygonAddedToManager) {
                this._init();
            }
        };
        NgMapsPolygon.prototype.ngOnChanges = function (changes) {
            if (!this._polygonAddedToManager) {
                this._init();
                return;
            }
            this._polygonManager.setPolygonOptions(this, this._updatePolygonOptions(changes));
        };
        NgMapsPolygon.prototype._init = function () {
            this._polygonManager.addPolygon(this);
            this._polygonAddedToManager = true;
            this._addEventListeners();
        };
        NgMapsPolygon.prototype._addEventListeners = function () {
            var _this = this;
            var handlers = [
                {
                    name: 'click',
                    handler: function (ev) { return _this.polyClick.emit(ev); },
                },
                {
                    name: 'dblclick',
                    handler: function (ev) { return _this.polyDblClick.emit(ev); },
                },
                { name: 'drag', handler: function (ev) { return _this.polyDrag.emit(ev); } },
                {
                    name: 'dragend',
                    handler: function (ev) { return _this.polyDragEnd.emit(ev); },
                },
                {
                    name: 'dragstart',
                    handler: function (ev) { return _this.polyDragStart.emit(ev); },
                },
                {
                    name: 'mousedown',
                    handler: function (ev) { return _this.polyMouseDown.emit(ev); },
                },
                {
                    name: 'mousemove',
                    handler: function (ev) { return _this.polyMouseMove.emit(ev); },
                },
                {
                    name: 'mouseout',
                    handler: function (ev) { return _this.polyMouseOut.emit(ev); },
                },
                {
                    name: 'mouseover',
                    handler: function (ev) { return _this.polyMouseOver.emit(ev); },
                },
                {
                    name: 'mouseup',
                    handler: function (ev) { return _this.polyMouseUp.emit(ev); },
                },
                {
                    name: 'rightclick',
                    handler: function (ev) { return _this.polyRightClick.emit(ev); },
                },
            ];
            handlers.forEach(function (obj) {
                var os = _this._polygonManager
                    .createEventObservable(obj.name, _this)
                    .subscribe(obj.handler);
                _this.subscription.add(os);
            });
        };
        NgMapsPolygon.prototype._updatePolygonOptions = function (changes) {
            return Object.keys(changes)
                .filter(function (k) { return NgMapsPolygon._polygonOptionsAttributes.indexOf(k) !== -1; })
                .reduce(function (obj, k) {
                obj[k] = changes[k].currentValue;
                return obj;
            }, {});
        };
        /** @internal */
        NgMapsPolygon.prototype.id = function () {
            return this._id;
        };
        /** @internal */
        NgMapsPolygon.prototype.ngOnDestroy = function () {
            this._polygonManager.deletePolygon(this);
            // unsubscribe all registered observable subscriptions
            this.subscription.unsubscribe();
        };
        return NgMapsPolygon;
    }());
    NgMapsPolygon._polygonOptionsAttributes = [
        'clickable',
        'draggable',
        'editable',
        'fillColor',
        'fillOpacity',
        'geodesic',
        'icon',
        'map',
        'paths',
        'strokeColor',
        'strokeOpacity',
        'strokeWeight',
        'visible',
        'zIndex',
        'draggable',
        'editable',
        'visible',
    ];
    NgMapsPolygon.decorators = [
        { type: i0.Directive, args: [{
                    selector: 'map-polygon',
                },] }
    ];
    NgMapsPolygon.ctorParameters = function () { return [
        { type: PolygonManager }
    ]; };
    NgMapsPolygon.propDecorators = {
        clickable: [{ type: i0.Input }],
        draggable: [{ type: i0.Input, args: ['polyDraggable',] }],
        editable: [{ type: i0.Input }],
        fillColor: [{ type: i0.Input }],
        fillOpacity: [{ type: i0.Input }],
        geodesic: [{ type: i0.Input }],
        paths: [{ type: i0.Input }],
        strokeColor: [{ type: i0.Input }],
        strokeOpacity: [{ type: i0.Input }],
        strokeWeight: [{ type: i0.Input }],
        visible: [{ type: i0.Input }],
        zIndex: [{ type: i0.Input }],
        polyClick: [{ type: i0.Output }],
        polyDblClick: [{ type: i0.Output }],
        polyDrag: [{ type: i0.Output }],
        polyDragEnd: [{ type: i0.Output }],
        polyDragStart: [{ type: i0.Output }],
        polyMouseDown: [{ type: i0.Output }],
        polyMouseMove: [{ type: i0.Output }],
        polyMouseOut: [{ type: i0.Output }],
        polyMouseOver: [{ type: i0.Output }],
        polyMouseUp: [{ type: i0.Output }],
        polyRightClick: [{ type: i0.Output }]
    };

    /**
     * NgMapsPolylinePoint represents one element of a polyline within a  {@link
     * NgMapsPolyline}
     */
    var NgMapsPolylinePoint = /** @class */ (function () {
        function NgMapsPolylinePoint() {
            /**
             * This event emitter gets emitted when the position of the point changed.
             */
            this.positionChanged = new i0.EventEmitter();
        }
        NgMapsPolylinePoint.prototype.ngOnChanges = function (changes) {
            if (changes.latitude || changes.longitude) {
                var position = {
                    lat: changes.latitude ? changes.latitude.currentValue : this.latitude,
                    lng: changes.longitude
                        ? changes.longitude.currentValue
                        : this.longitude,
                };
                this.positionChanged.emit(position);
            }
        };
        return NgMapsPolylinePoint;
    }());
    NgMapsPolylinePoint.decorators = [
        { type: i0.Directive, args: [{
                    selector: 'map-polyline-point',
                },] }
    ];
    NgMapsPolylinePoint.ctorParameters = function () { return []; };
    NgMapsPolylinePoint.propDecorators = {
        latitude: [{ type: i0.Input }],
        longitude: [{ type: i0.Input }],
        positionChanged: [{ type: i0.Output }]
    };

    var PolylineManager = /** @class */ (function () {
        function PolylineManager(_mapsWrapper, _zone) {
            this._mapsWrapper = _mapsWrapper;
            this._zone = _zone;
            this._polylines = new Map();
        }
        PolylineManager.prototype._convertPoints = function (line) {
            return line._getPoints().map(function (point) {
                return {
                    lat: point.latitude,
                    lng: point.longitude,
                };
            });
        };
        return PolylineManager;
    }());
    PolylineManager.decorators = [
        { type: i0.Injectable }
    ];
    PolylineManager.ctorParameters = function () { return [
        { type: MapsApiWrapper },
        { type: i0.NgZone }
    ]; };

    var polylineId = 0;
    /**
     * NgMapsPolyline renders a polyline on a {@link https://ng-maps.github.io/core/components/NgMapsViewComponent.html|NgMapsView}
     */
    var NgMapsPolyline = /** @class */ (function () {
        function NgMapsPolyline(_polylineManager) {
            this._polylineManager = _polylineManager;
            /**
             * Indicates whether this Polyline handles mouse events. Defaults to true.
             */
            this.clickable = true;
            /**
             * If set to true, the user can drag this shape over the map. The geodesic property defines the
             * mode of dragging. Defaults to false.
             */
            // tslint:disable-next-line:no-input-rename
            this.draggable = false;
            /**
             * If set to true, the user can edit this shape by dragging the control points shown at the
             * vertices and on each segment. Defaults to false.
             */
            this.editable = false;
            /**
             * When true, edges of the polygon are interpreted as geodesic and will follow the curvature of
             * the Earth. When false, edges of the polygon are rendered as straight lines in screen space.
             * Note that the shape of a geodesic polygon may appear to change when dragged, as the dimensions
             * are maintained relative to the surface of the earth. Defaults to false.
             */
            this.geodesic = false;
            /**
             * Whether this polyline is visible on the map. Defaults to true.
             */
            this.visible = true;
            /**
             * This event is fired when the DOM click event is fired on the Polyline.
             */
            this.lineClick = new i0.EventEmitter();
            /**
             * This event is fired when the DOM dblclick event is fired on the Polyline.
             */
            this.lineDblClick = new i0.EventEmitter();
            /**
             * This event is repeatedly fired while the user drags the polyline.
             */
            this.lineDrag = new i0.EventEmitter();
            /**
             * This event is fired when the user stops dragging the polyline.
             */
            this.lineDragEnd = new i0.EventEmitter();
            /**
             * This event is fired when the user starts dragging the polyline.
             */
            this.lineDragStart = new i0.EventEmitter();
            /**
             * This event is fired when the DOM mousedown event is fired on the Polyline.
             */
            this.lineMouseDown = new i0.EventEmitter();
            /**
             * This event is fired when the DOM mousemove event is fired on the Polyline.
             */
            this.lineMouseMove = new i0.EventEmitter();
            /**
             * This event is fired on Polyline mouseout.
             */
            this.lineMouseOut = new i0.EventEmitter();
            /**
             * This event is fired on Polyline mouseover.
             */
            this.lineMouseOver = new i0.EventEmitter();
            /**
             * This event is fired whe the DOM mouseup event is fired on the Polyline
             */
            this.lineMouseUp = new i0.EventEmitter();
            /**
             * This even is fired when the Polyline is right-clicked on.
             */
            this.lineRightClick = new i0.EventEmitter();
            this._polylineAddedToManager = false;
            this.subscription = new rxjs.Subscription();
            this._id = (polylineId++).toString();
        }
        /** @internal */
        NgMapsPolyline.prototype.ngAfterContentInit = function () {
            var _this = this;
            if (this.points.length) {
                this.points.forEach(function (point) {
                    var subscription = point.positionChanged.subscribe(function () {
                        _this._polylineManager.updatePolylinePoints(_this);
                    });
                    _this.subscription.add(subscription);
                });
            }
            if (!this._polylineAddedToManager) {
                this._init();
            }
            var s = this.points.changes.subscribe(function () { return _this._polylineManager.updatePolylinePoints(_this); });
            this.subscription.add(s);
            this._polylineManager.updatePolylinePoints(this);
        };
        NgMapsPolyline.prototype.ngOnChanges = function (changes) {
            if (!this._polylineAddedToManager) {
                this._init();
                return;
            }
            var options = {};
            var optionKeys = Object.keys(changes).filter(function (k) { return NgMapsPolyline._polylineOptionsAttributes.indexOf(k) !== -1; });
            optionKeys.forEach(function (k) { return (options[k] = changes[k].currentValue); });
            this._polylineManager.setPolylineOptions(this, options);
        };
        NgMapsPolyline.prototype._init = function () {
            this._polylineManager.addPolyline(this);
            this._polylineAddedToManager = true;
            this._addEventListeners();
        };
        NgMapsPolyline.prototype._addEventListeners = function () {
            var _this = this;
            var handlers = [
                {
                    name: 'click',
                    handler: function (ev) { return _this.lineClick.emit(ev); },
                },
                {
                    name: 'dblclick',
                    handler: function (ev) { return _this.lineDblClick.emit(ev); },
                },
                { name: 'drag', handler: function (ev) { return _this.lineDrag.emit(ev); } },
                {
                    name: 'dragend',
                    handler: function (ev) { return _this.lineDragEnd.emit(ev); },
                },
                {
                    name: 'dragstart',
                    handler: function (ev) { return _this.lineDragStart.emit(ev); },
                },
                {
                    name: 'mousedown',
                    handler: function (ev) { return _this.lineMouseDown.emit(ev); },
                },
                {
                    name: 'mousemove',
                    handler: function (ev) { return _this.lineMouseMove.emit(ev); },
                },
                {
                    name: 'mouseout',
                    handler: function (ev) { return _this.lineMouseOut.emit(ev); },
                },
                {
                    name: 'mouseover',
                    handler: function (ev) { return _this.lineMouseOver.emit(ev); },
                },
                {
                    name: 'mouseup',
                    handler: function (ev) { return _this.lineMouseUp.emit(ev); },
                },
                {
                    name: 'rightclick',
                    handler: function (ev) { return _this.lineRightClick.emit(ev); },
                },
            ];
            handlers.forEach(function (obj) {
                var os = _this._polylineManager
                    .createEventObservable(obj.name, _this)
                    .subscribe(obj.handler);
                _this.subscription.add(os);
            });
        };
        /** @internal */
        NgMapsPolyline.prototype._getPoints = function () {
            if (this.points) {
                return this.points.toArray();
            }
            else {
                return [];
            }
        };
        /** @internal */
        NgMapsPolyline.prototype.id = function () {
            return this._id;
        };
        /** @internal */
        NgMapsPolyline.prototype.ngOnDestroy = function () {
            this._polylineManager.deletePolyline(this);
            // unsubscribe all registered observable subscriptions
            this.subscription.unsubscribe();
        };
        return NgMapsPolyline;
    }());
    NgMapsPolyline._polylineOptionsAttributes = [
        'draggable',
        'editable',
        'visible',
        'geodesic',
        'strokeColor',
        'strokeOpacity',
        'strokeWeight',
        'zIndex',
        'icons',
    ];
    NgMapsPolyline.decorators = [
        { type: i0.Directive, args: [{
                    selector: 'map-polyline',
                },] }
    ];
    NgMapsPolyline.ctorParameters = function () { return [
        { type: PolylineManager }
    ]; };
    NgMapsPolyline.propDecorators = {
        clickable: [{ type: i0.Input }],
        draggable: [{ type: i0.Input, args: ['polylineDraggable',] }],
        editable: [{ type: i0.Input }],
        geodesic: [{ type: i0.Input }],
        strokeColor: [{ type: i0.Input }],
        strokeOpacity: [{ type: i0.Input }],
        strokeWeight: [{ type: i0.Input }],
        icons: [{ type: i0.Input }],
        visible: [{ type: i0.Input }],
        zIndex: [{ type: i0.Input }],
        lineClick: [{ type: i0.Output }],
        lineDblClick: [{ type: i0.Output }],
        lineDrag: [{ type: i0.Output }],
        lineDragEnd: [{ type: i0.Output }],
        lineDragStart: [{ type: i0.Output }],
        lineMouseDown: [{ type: i0.Output }],
        lineMouseMove: [{ type: i0.Output }],
        lineMouseOut: [{ type: i0.Output }],
        lineMouseOver: [{ type: i0.Output }],
        lineMouseUp: [{ type: i0.Output }],
        lineRightClick: [{ type: i0.Output }],
        points: [{ type: i0.ContentChildren, args: [NgMapsPolylinePoint,] }]
    };

    var RectangleManager = /** @class */ (function () {
        function RectangleManager(_apiWrapper, _zone) {
            this._apiWrapper = _apiWrapper;
            this._zone = _zone;
            this._rectangles = new Map();
        }
        return RectangleManager;
    }());
    RectangleManager.decorators = [
        { type: i0.Injectable }
    ];
    RectangleManager.ctorParameters = function () { return [
        { type: MapsApiWrapper },
        { type: i0.NgZone }
    ]; };

    var NgMapsRectangleDirective = /** @class */ (function () {
        function NgMapsRectangleDirective(_manager) {
            this._manager = _manager;
            /**
             * Indicates whether this Rectangle handles mouse events. Defaults to true.
             */
            this.clickable = true;
            /**
             * If set to true, the user can drag this rectangle over the map. Defaults to false.
             */
            // tslint:disable-next-line:no-input-rename
            this.draggable = false;
            /**
             * If set to true, the user can edit this rectangle by dragging the control points shown at
             * the center and around the circumference of the rectangle. Defaults to false.
             */
            this.editable = false;
            /**
             * The stroke position. Defaults to CENTER.
             * This property is not supported on Internet Explorer 8 and earlier.
             */
            this.strokePosition = 'CENTER';
            /**
             * The stroke width in pixels.
             */
            this.strokeWeight = 0;
            /**
             * Whether this rectangle is visible on the map. Defaults to true.
             */
            this.visible = true;
            /**
             * This event is fired when the rectangle's is changed.
             */
            this.boundsChange = new i0.EventEmitter();
            /**
             * This event emitter gets emitted when the user clicks on the rectangle.
             */
            this.rectangleClick = new i0.EventEmitter();
            /**
             * This event emitter gets emitted when the user clicks on the rectangle.
             */
            this.rectangleDblClick = new i0.EventEmitter();
            /**
             * This event is repeatedly fired while the user drags the rectangle.
             */
            this.drag = new i0.EventEmitter();
            /**
             * This event is fired when the user stops dragging the rectangle.
             */
            this.dragEnd = new i0.EventEmitter();
            /**
             * This event is fired when the user starts dragging the rectangle.
             */
            this.dragStart = new i0.EventEmitter();
            /**
             * This event is fired when the DOM mousedown event is fired on the rectangle.
             */
            this.mouseDown = new i0.EventEmitter();
            /**
             * This event is fired when the DOM mousemove event is fired on the rectangle.
             */
            this.mouseMove = new i0.EventEmitter();
            /**
             * This event is fired on rectangle mouseout.
             */
            this.mouseOut = new i0.EventEmitter();
            /**
             * This event is fired on rectangle mouseover.
             */
            this.mouseOver = new i0.EventEmitter();
            /**
             * This event is fired when the DOM mouseup event is fired on the rectangle.
             */
            this.mouseUp = new i0.EventEmitter();
            /**
             * This event is fired when the rectangle is right-clicked on.
             */
            this.rightClick = new i0.EventEmitter();
            this._rectangleAddedToManager = false;
            this.subscription = new rxjs.Subscription();
        }
        /** @internal */
        NgMapsRectangleDirective.prototype.ngOnInit = function () {
            this._manager.addRectangle(this);
            this._rectangleAddedToManager = true;
            this._registerEventListeners();
        };
        /** @internal */
        NgMapsRectangleDirective.prototype.ngOnChanges = function (changes) {
            if (!this._rectangleAddedToManager) {
                return;
            }
            if (changes.north || changes.east || changes.south || changes.west) {
                this._manager.setBounds(this);
            }
            if (changes.editable) {
                this._manager.setEditable(this);
            }
            if (changes.draggable) {
                this._manager.setDraggable(this);
            }
            if (changes.visible) {
                this._manager.setVisible(this);
            }
            this._updateRectangleOptionsChanges(changes);
        };
        NgMapsRectangleDirective.prototype._updateRectangleOptionsChanges = function (changes) {
            var options = {};
            var optionKeys = Object.keys(changes).filter(function (k) { return NgMapsRectangleDirective._mapOptions.indexOf(k) !== -1; });
            optionKeys.forEach(function (k) {
                options[k] = changes[k].currentValue;
            });
            if (optionKeys.length > 0) {
                this._manager.setOptions(this, options);
            }
        };
        NgMapsRectangleDirective.prototype._registerEventListeners = function () {
            var _this = this;
            var events = new Map();
            events.set('bounds_changed', this.boundsChange);
            events.set('click', this.rectangleClick);
            events.set('dblclick', this.rectangleDblClick);
            events.set('drag', this.drag);
            events.set('dragend', this.dragEnd);
            events.set('dragStart', this.dragStart);
            events.set('mousedown', this.mouseDown);
            events.set('mousemove', this.mouseMove);
            events.set('mouseout', this.mouseOut);
            events.set('mouseover', this.mouseOver);
            events.set('mouseup', this.mouseUp);
            events.set('rightclick', this.rightClick);
            events.forEach(function (eventEmitter, eventName) {
                _this.subscription.add(_this._manager
                    .createEventObservable(eventName, _this)
                    .subscribe(function (value) {
                    switch (eventName) {
                        case 'bounds_changed':
                            _this._manager
                                .getBounds(_this)
                                .then(function (bounds) { return eventEmitter.emit(bounds); });
                            break;
                        default:
                            eventEmitter.emit(value);
                    }
                }));
            });
        };
        /** @internal */
        NgMapsRectangleDirective.prototype.ngOnDestroy = function () {
            this.subscription.unsubscribe();
            this._manager.removeRectangle(this);
        };
        /**
         * Gets the LatLngBounds of this Rectangle.
         */
        NgMapsRectangleDirective.prototype.getBounds = function () {
            return this._manager.getBounds(this);
        };
        return NgMapsRectangleDirective;
    }());
    NgMapsRectangleDirective._mapOptions = [
        'fillColor',
        'fillOpacity',
        'strokeColor',
        'strokeOpacity',
        'strokePosition',
        'strokeWeight',
        'visible',
        'zIndex',
        'clickable',
    ];
    NgMapsRectangleDirective.decorators = [
        { type: i0.Directive, args: [{
                    selector: 'map-rectangle',
                },] }
    ];
    NgMapsRectangleDirective.ctorParameters = function () { return [
        { type: RectangleManager }
    ]; };
    NgMapsRectangleDirective.propDecorators = {
        north: [{ type: i0.Input }],
        east: [{ type: i0.Input }],
        south: [{ type: i0.Input }],
        west: [{ type: i0.Input }],
        clickable: [{ type: i0.Input }],
        draggable: [{ type: i0.Input, args: ['rectangleDraggable',] }],
        editable: [{ type: i0.Input }],
        fillColor: [{ type: i0.Input }],
        fillOpacity: [{ type: i0.Input }],
        strokeColor: [{ type: i0.Input }],
        strokeOpacity: [{ type: i0.Input }],
        strokePosition: [{ type: i0.Input }],
        strokeWeight: [{ type: i0.Input }],
        visible: [{ type: i0.Input }],
        zIndex: [{ type: i0.Input }],
        boundsChange: [{ type: i0.Output }],
        rectangleClick: [{ type: i0.Output }],
        rectangleDblClick: [{ type: i0.Output }],
        drag: [{ type: i0.Output }],
        dragEnd: [{ type: i0.Output }],
        dragStart: [{ type: i0.Output }],
        mouseDown: [{ type: i0.Output }],
        mouseMove: [{ type: i0.Output }],
        mouseOut: [{ type: i0.Output }],
        mouseOver: [{ type: i0.Output }],
        mouseUp: [{ type: i0.Output }],
        rightClick: [{ type: i0.Output }]
    };

    /**
     * When using the NoOpMapsAPILoader, the Google Maps API must be added to the page via a `<script>`
     * Tag.
     * It's important that the Google Maps API script gets loaded first on the page.
     */
    var NoOpMapsAPILoader = /** @class */ (function (_super) {
        __extends(NoOpMapsAPILoader, _super);
        function NoOpMapsAPILoader(document) {
            var _this = _super.call(this) || this;
            _this._document = document;
            _this._window = _this._document.defaultView;
            return _this;
        }
        NoOpMapsAPILoader.prototype.load = function () {
            return Promise.resolve();
        };
        NoOpMapsAPILoader.prototype.configure = function (config) { };
        return NoOpMapsAPILoader;
    }(MapsAPILoader));
    NoOpMapsAPILoader.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: i0.Inject, args: [i1.DOCUMENT,] }] }
    ]; };

    var ScriptLoaderService = /** @class */ (function () {
        function ScriptLoaderService(document) {
            this.alreadyLoaded = new Map();
            this.document = document;
            this.head = this.document.head;
        }
        ScriptLoaderService.prototype.observeLoad = function (element) {
            return new Promise(function (resolve, reject) {
                function successHandler(event) {
                    var readyState = element.readyState;
                    // For IE we have readyState, other browsers just call the load event and we proccede
                    if (readyState === 'complete' ||
                        readyState === 'loaded' ||
                        event.type === 'load') {
                        resolve(event);
                    }
                }
                function errorHandler(event) {
                    reject(event);
                }
                element.addEventListener('readystatechange', successHandler);
                element.addEventListener('load', successHandler);
                element.addEventListener('error', errorHandler);
            });
        };
        ScriptLoaderService.prototype.createScriptElement = function (src) {
            var script = document.createElement('script');
            script.src = src;
            script.type = 'text/javascript';
            script.async = true;
            script.defer = true;
            return script;
        };
        // tslint:disable-next-line:naming-convention
        ScriptLoaderService.prototype.createCSSElement = function (href) {
            var style = document.createElement('link');
            style.rel = 'stylesheet';
            style.type = 'text/css';
            style.href = href;
            return style;
        };
        /**
         * Dynamically loads the given script
         * @param src The url of the script to load dynamically
         * @param integrity set integrity check value
         * @returns Observable<UIEvent> Observable that will be resolved once the script has been loaded.
         */
        ScriptLoaderService.prototype.loadScript = function (src, integrity) {
            if (this.alreadyLoaded.has(src)) {
                return this.alreadyLoaded.get(src);
            }
            else {
                var script = this.createScriptElement(src);
                var promise = this.observeLoad(script);
                this.head.appendChild(script);
                this.alreadyLoaded.set(src, promise);
                return promise;
            }
        };
        /**
         * Dynamically loads the given CSS file
         * @param href The url of the CSS to load dynamically
         * @returns Observable<UIEvent> Promise that will be resolved once the CSS file has been loaded.
         */
        ScriptLoaderService.prototype.loadCSS = function (href) {
            // tslint:disable-line:naming-convention
            if (this.alreadyLoaded.has(href)) {
                return this.alreadyLoaded.get(href);
            }
            else {
                var style = this.createCSSElement(href);
                var promise = this.observeLoad(style);
                this.head.appendChild(style);
                this.alreadyLoaded.set(href, promise);
                return promise;
            }
        };
        return ScriptLoaderService;
    }());
    ScriptLoaderService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ScriptLoaderService_Factory() { return new ScriptLoaderService(i0.ɵɵinject(i1.DOCUMENT)); }, token: ScriptLoaderService, providedIn: "root" });
    ScriptLoaderService.decorators = [
        { type: i0.Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    ScriptLoaderService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: i0.Inject, args: [i1.DOCUMENT,] }] }
    ]; };

    var MAP_PROVIDER = new i0.InjectionToken('Map Provider');
    function isGoogleMaps(token) {
        return token === 'GoogleMaps';
    }
    function isHereMaps(token) {
        return token === 'HereMaps';
    }
    /**
     * @internal
     */
    function coreDirectives() {
        return [
            NgMapsViewComponent,
            NgMapsMarkerComponent,
            NgMapsInfoWindowComponent,
            NgMapsFitBoundsDirective,
            NgMapsCircleDirective,
            NgMapsPolygon,
            NgMapsPolyline,
            NgMapsPolylinePoint,
            NgMapsRectangleDirective,
        ];
    }
    /**
     * The ng-maps core module. Contains all Directives/Services/Pipes
     */
    var NgMapsCoreModule = /** @class */ (function () {
        function NgMapsCoreModule() {
        }
        return NgMapsCoreModule;
    }());
    NgMapsCoreModule.decorators = [
        { type: i0.NgModule, args: [{
                    declarations: coreDirectives(),
                    exports: coreDirectives(),
                },] }
    ];

    /*
     * Public API Surface of core
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.CircleManager = CircleManager;
    exports.FitBoundsAccessor = FitBoundsAccessor;
    exports.FitBoundsService = FitBoundsService;
    exports.InfoWindowManager = InfoWindowManager;
    exports.MAP_PROVIDER = MAP_PROVIDER;
    exports.MapsAPILoader = MapsAPILoader;
    exports.MapsApiWrapper = MapsApiWrapper;
    exports.MarkerManager = MarkerManager;
    exports.NgMapsCircleDirective = NgMapsCircleDirective;
    exports.NgMapsCoreModule = NgMapsCoreModule;
    exports.NgMapsFitBoundsDirective = NgMapsFitBoundsDirective;
    exports.NgMapsInfoWindowComponent = NgMapsInfoWindowComponent;
    exports.NgMapsMarkerComponent = NgMapsMarkerComponent;
    exports.NgMapsPolygon = NgMapsPolygon;
    exports.NgMapsPolyline = NgMapsPolyline;
    exports.NgMapsPolylinePoint = NgMapsPolylinePoint;
    exports.NgMapsRectangleDirective = NgMapsRectangleDirective;
    exports.NgMapsViewComponent = NgMapsViewComponent;
    exports.NoOpMapsAPILoader = NoOpMapsAPILoader;
    exports.PolygonManager = PolygonManager;
    exports.PolylineManager = PolylineManager;
    exports.RectangleManager = RectangleManager;
    exports.ScriptLoaderService = ScriptLoaderService;
    exports.isGoogleMaps = isGoogleMaps;
    exports.isHereMaps = isHereMaps;
    exports.ɵa = coreDirectives;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
